{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Robustness and Plasticity in Regulatory Networks\n",
    "\n",
    "* [Introduction](./RPRN-Introduction.ipynb)\n",
    "* [BoolNet](#BoolNet)\n",
    "    * [Network construction](#Network-construction)\n",
    "    * [Attractors](#Attractors)\n",
    "    * [Labeling](#Labeling)\n",
    "* [Functions](./RPRN-Functions.ipynb)\n",
    "* [Updating](./RPRN-Updating.ipynb)\n",
    "* [States](./RPRN-States.ipynb)\n",
    "* [Appendix](./RPRN-Appendix.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BoolNet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once a regulatory network has been constructed there exist [multiple tools](http://www.colomoto.org/software/) to analyse it. In this tutorial we will focus in the R package [Boolnet](https://cran.r-project.org/web/packages/BoolNet/index.html).\n",
    "\n",
    "It is possible to combine the different software tools to make a better analysis, the standar format for saving a logical regulatory network is [SBML-qual](http://www.colomoto.org/formats/sbml-qual.html) (see the [Appendix](./RPRN-Appendix.ipynb) for more information).\n",
    "\n",
    "Sometimes it is useful to [simplify the network](http://ginsim.org/ginsim-doc/current/algo-reduction.html). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Network construction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The interactions of a biological system can be expressed as a set of logical functions using multiple formalisms ([Boolean functions](https://en.wikipedia.org/wiki/Boolean_function), [truth tables](https://en.wikipedia.org/wiki/Truth_table), [Binary Decision Diagrams](https://en.wikipedia.org/wiki/Binary_decision_diagram)). BoolNet accepts a set of boolean functions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For examplet the Th17/iTreg network can be expressed as a set of boolean functions obtained from the known interactions among the cytokines and transcription factors.\n",
    "\n",
    ">targets, functions\n",
    "\n",
    ">IL2, (IL2e | (IL2 &  ! FOXP3)) &  ! (STAT3 | (IL10 & ! FOXP3))\n",
    "\n",
    ">RORGT, (STAT3 & TGFB) &  ! FOXP3\n",
    "\n",
    ">STAT3, (IL21e | STAT3 | RORGT) &  ! (IL10 | IL2)\n",
    "\n",
    ">FOXP3, (IL2 & (TGFB | FOXP3)) &  ! (STAT3 | RORGT)\n",
    "\n",
    ">TGFB, TGFBe | ((TGFB | FOXP3) &  ! STAT3 )\n",
    "\n",
    ">IL10, IL10e | (IL10 & (STAT3 | TGFB))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we will load BoolNet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Uncomment next line if you haven't installed BoolNet\n",
    "#install.packages(\"BoolNet\", repos='http://cran.us.r-project.org')\n",
    "library(BoolNet)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Boolnet creates a network object with the function __loadNetwork()__. This function accepts the filename of the file with the rules. In this case we will create the file from the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fileConn<-file(\"minTh17iTreg.txt\")\n",
    "writeLines(c(\n",
    "    \"targets, functions\",\n",
    "    \"IL2, (IL2e | (IL2 &  ! FOXP3)) &  ! (STAT3 | (IL10 & ! FOXP3))\",\n",
    "    \"RORGT, (STAT3 & TGFB) &  ! FOXP3\",\n",
    "    \"STAT3, (IL21e | STAT3 | RORGT) &  ! (IL10 | IL2)\",\n",
    "    \"FOXP3, (IL2 & (TGFB | FOXP3)) &  ! (STAT3 | RORGT)\",\n",
    "    \"TGFB, TGFBe | ((TGFB | FOXP3) &  ! STAT3 )\",\n",
    "    \"IL10, IL10e | (IL10 & (STAT3 | TGFB))\", \n",
    "    \"IL2e, IL2e\",\n",
    "    \"IL21e, IL21e\",\n",
    "    \"TGFBe, TGFBe\",\n",
    "    \"IL10e, IL10e\"\n",
    "    ), fileConn)\n",
    "close(fileConn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can create a network object. If you don't specify a node BoolNet will raise a warning and asumme it is an input. The resulting network should have ten nodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Boolean network with 10 genes\n",
       "\n",
       "Involved genes:\n",
       "IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Transition functions:\n",
       "IL2 = (IL2e | (IL2 &  ! FOXP3)) &  ! (STAT3 | (IL10 & ! FOXP3))\n",
       "RORGT = (STAT3 & TGFB) &  ! FOXP3\n",
       "STAT3 = (IL21e | STAT3 | RORGT) &  ! (IL10 | IL2)\n",
       "FOXP3 = (IL2 & (TGFB | FOXP3)) &  ! (STAT3 | RORGT)\n",
       "TGFB = TGFBe | ((TGFB | FOXP3) &  ! STAT3 )\n",
       "IL10 = IL10e | (IL10 & (STAT3 | TGFB))\n",
       "IL2e = IL2e\n",
       "IL21e = IL21e\n",
       "TGFBe = TGFBe\n",
       "IL10e = IL10e"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "net <- loadNetwork(\"minTh17iTreg.txt\")\n",
    "net"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The result is a network object. This object structure is:\n",
    "* __network__\n",
    "    * __\\$interactions__ : list of nodes where each node has its functions represented as truth tables. (Use iterator or [[n]] to examine each node)\n",
    "        * __\\$input__ : index of regulators\n",
    "        * __\\$func__ : function as truth table\n",
    "        * __\\$expression__ : Boolean function as string\n",
    "    * __\\$genes__ : ordered node names\n",
    "    * __\\$fixed__ : which nodes have fixed values (see [Functions](./RPRN-Functions.ipynb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "$input\n",
       "[1] 1 3 4 6 7\n",
       "\n",
       "$func\n",
       " [1] 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 1 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0\n",
       "\n",
       "$expression\n",
       "[1] \"(IL2e | (IL2 &  ! FOXP3)) &  ! (STAT3 | (IL10 & ! FOXP3))\"\n"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'IL2'</li>\n",
       "\t<li>'RORGT'</li>\n",
       "\t<li>'STAT3'</li>\n",
       "\t<li>'FOXP3'</li>\n",
       "\t<li>'TGFB'</li>\n",
       "\t<li>'IL10'</li>\n",
       "\t<li>'IL2e'</li>\n",
       "\t<li>'IL21e'</li>\n",
       "\t<li>'TGFBe'</li>\n",
       "\t<li>'IL10e'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'IL2'\n",
       "\\item 'RORGT'\n",
       "\\item 'STAT3'\n",
       "\\item 'FOXP3'\n",
       "\\item 'TGFB'\n",
       "\\item 'IL10'\n",
       "\\item 'IL2e'\n",
       "\\item 'IL21e'\n",
       "\\item 'TGFBe'\n",
       "\\item 'IL10e'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'IL2'\n",
       "2. 'RORGT'\n",
       "3. 'STAT3'\n",
       "4. 'FOXP3'\n",
       "5. 'TGFB'\n",
       "6. 'IL10'\n",
       "7. 'IL2e'\n",
       "8. 'IL21e'\n",
       "9. 'TGFBe'\n",
       "10. 'IL10e'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"IL2\"   \"RORGT\" \"STAT3\" \"FOXP3\" \"TGFB\"  \"IL10\"  \"IL2e\"  \"IL21e\" \"TGFBe\"\n",
       "[10] \"IL10e\""
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<dl class=dl-horizontal>\n",
       "\t<dt>IL2</dt>\n",
       "\t\t<dd>-1</dd>\n",
       "\t<dt>RORGT</dt>\n",
       "\t\t<dd>-1</dd>\n",
       "\t<dt>STAT3</dt>\n",
       "\t\t<dd>-1</dd>\n",
       "\t<dt>FOXP3</dt>\n",
       "\t\t<dd>-1</dd>\n",
       "\t<dt>TGFB</dt>\n",
       "\t\t<dd>-1</dd>\n",
       "\t<dt>IL10</dt>\n",
       "\t\t<dd>-1</dd>\n",
       "\t<dt>IL2e</dt>\n",
       "\t\t<dd>-1</dd>\n",
       "\t<dt>IL21e</dt>\n",
       "\t\t<dd>-1</dd>\n",
       "\t<dt>TGFBe</dt>\n",
       "\t\t<dd>-1</dd>\n",
       "\t<dt>IL10e</dt>\n",
       "\t\t<dd>-1</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[IL2] -1\n",
       "\\item[RORGT] -1\n",
       "\\item[STAT3] -1\n",
       "\\item[FOXP3] -1\n",
       "\\item[TGFB] -1\n",
       "\\item[IL10] -1\n",
       "\\item[IL2e] -1\n",
       "\\item[IL21e] -1\n",
       "\\item[TGFBe] -1\n",
       "\\item[IL10e] -1\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "IL2\n",
       ":   -1RORGT\n",
       ":   -1STAT3\n",
       ":   -1FOXP3\n",
       ":   -1TGFB\n",
       ":   -1IL10\n",
       ":   -1IL2e\n",
       ":   -1IL21e\n",
       ":   -1TGFBe\n",
       ":   -1IL10e\n",
       ":   -1\n",
       "\n"
      ],
      "text/plain": [
       "  IL2 RORGT STAT3 FOXP3  TGFB  IL10  IL2e IL21e TGFBe IL10e \n",
       "   -1    -1    -1    -1    -1    -1    -1    -1    -1    -1 "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "net$interactions[[1]]\n",
    "net$genes\n",
    "net$fixed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Attractors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[//]: # (As the state of the network is updated using the functions, the network will reach a previously visited state called an attractor. The set of states that lead to an attractor is called the basin of the attractor [Fig1 C] \\cite{Gerherson2004}.)\n",
    "\n",
    "Attractors represent stable states in the dynamics of the network and have been related to cell types or  biological processes like the cell cycle. \n",
    "\n",
    "The number of possible states and the complexity of finding the attractors grows exponentially with the number of nodes. It is possible efficiently find the stable states [without simulations](http://ginsim.org/documentation) and using [model checking](http://dl.acm.org/citation.cfm?id=2014689). \n",
    "\n",
    "The attractors we obtain can be affected by the [updating method](./RPRN-Updating.ipynb) (synchronus or asynchronous)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "BoolNet uses the function __getAttractors()__ to obtain the attractors of the network. This function has a lot of options that can be checked in the documentation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for getAttractors {BoolNet}\"><tr><td>getAttractors {BoolNet}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Identify attractors in a Boolean network</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Identifies attractors (cycles) in a supplied Boolean network using synchronous or asynchronous state transitions</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "getAttractors(network, \n",
       "              type = c(\"synchronous\",\"asynchronous\"), \n",
       "              method = c(\"exhaustive\",\n",
       "                         \"sat.exhaustive\",\n",
       "                         \"sat.restricted\",\n",
       "                         \"random\",\n",
       "                         \"chosen\"), \n",
       "              startStates = list(),\n",
       "              genesON = c(), genesOFF = c(), \n",
       "              canonical = TRUE,\n",
       "              randomChainLength = 10000, \n",
       "              avoidSelfLoops = TRUE, \n",
       "              geneProbabilities = NULL, \n",
       "              maxAttractorLength = Inf,\n",
       "              returnTable = TRUE) \n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>network</code></td>\n",
       "<td>\n",
       "\n",
       "<p>A network structure of class <code>BooleanNetwork</code> or <code>SymbolicBooleanNetwork</code>. These networks can be read from files by <code>loadNetwork</code>, generated by <code>generateRandomNKNetwork</code>, or reconstructed by <code>reconstructNetwork</code>.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>type</code></td>\n",
       "<td>\n",
       "\n",
       "<p>If <code>type=\"synchronous\"</code>, synchronous state transitions are used, i.e. all genes are updated at the same time. Synchronous attractor search can be performed in an exhaustive manner or using a heuristic that starts from predefined states. For symbolic networks, only synchronous updates are possible.\n",
       "</p>\n",
       "<p>If <code>type=\"asynchronous\"</code>, asynchronous state transitions are performed, i.e. one (randomly chosen) gene is updated in each transition. Steady-state attractors are the same in asynchronous and synchronous networks, but the asynchronous search is also able to identify complex/loose attractors. Asynchronous search relies on a heuristic algorithm that starts from predefined states.\n",
       "</p>\n",
       "<p>See Details for more information on the algorithms.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>method</code></td>\n",
       "<td>\n",
       "\n",
       "<p>The search method to be used. If &quot;exhaustive&quot;, attractors are identified by exhaustive state space search, i.e. by calculating the sucessors of all 2^n states (where n is the number of genes that are not set to a fixed value). This kind of search is only available for synchronous attractor search, and the maximum number of genes allowed for exhaustive search is 29. Apart from the attractors, this method generates the full state transition graph.\n",
       "</p>\n",
       "<p>If <code>method</code> is &quot;sat.exhaustive&quot; or &quot;sat.restricted&quot;, attractors are identified using algorithms based on the satisfiability problem. This search type is also restricted to synchronous networks. It can be used to identify attractors in much larger networks than with <code>method=\"exhaustive\"</code>, but does not return the state transition graph. For <code>method=\"sat.exhaustive\"</code>, an exhaustive attractor search is performed, while <code>method=\"sat.restricted\"</code> only searches for attractors of a specified maximum length <code>maxAttractorLength</code>.\n",
       "</p>\n",
       "<p>If <code>method</code> is &quot;random&quot;, <code>startStates</code> is interpreted as an integer value specifying the number of states to be generated randomly. The algorithm is then initialized with these random states and identifies the attractors to which these states lead.\n",
       "</p>\n",
       "<p>If <code>method</code> is &quot;chosen&quot;, <code>startStates</code> is interpreted as a list of binary vectors, each specifying one input state. Each vector must have <code>length(network$genes)</code> elements with 0 or 1 values. The algorithm identifies the attractors to which the supplied states lead. If <code>network</code> is of class <code>SymbolicBooleanNetwork</code> and makes use of more than one predecessor state, this can also be a list of matrices with the genes in the columns and multiple predecessor states in the rows.\n",
       "</p>\n",
       "<p>If <code>method</code> is not supplied, the desired method is inferred from the type of <code>startStates</code>. By default, if neither <code>method</code> nor <code>startStates</code> are provided, an exhaustive search is performed.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>startStates</code></td>\n",
       "<td>\n",
       "\n",
       "<p>The value of <code>startStates</code> depends on the chosen method. See <code>method</code> for more details.\n",
       "</p>\n",
       "</td></tr>  \n",
       "<tr valign=\"top\"><td><code>genesON</code></td>\n",
       "<td>\n",
       "\n",
       "<p>A vector of genes whose values are fixed to 1, which reduces the complexity of the search. This is equivalent to a preceding call of <code>fixGenes</code>.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>genesOFF</code></td>\n",
       "<td>\n",
       "\n",
       "<p>A vector of genes whose values are fixed to 0, which reduces the complexity of the search. This is equivalent to a preceding call of <code>fixGenes</code>.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>canonical</code></td>\n",
       "<td>\n",
       "\n",
       "<p>If set to true, the states in the attractors are rearranged such that the state whose binary encoding\n",
       "makes up the smallest number is the first element of the vector. This ensures that attractors found by different heuristic runs of <code>getAttractors</code> are comparable, as the cycles may have been entered at different states in different runs of the algorithm.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>randomChainLength</code></td>\n",
       "<td>\n",
       "\n",
       "<p>If <code>type=\"asynchronous\"</code>, this parameter specifies the number of random transitions performed by the search to enter a potential attractor (see Details). Defaults to 10000.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>avoidSelfLoops</code></td>\n",
       "<td>\n",
       "\n",
       "<p>If <code>type=\"asynchronous\"</code> and <code>avoidSelfLoops=TRUE</code>, the asynchronous attractor search only enters self loops (i.e. transitions that result in the same state) if none of the possible transitions can leave the state. This results in attractors with fewer edges. Otherwise, self loops are included in the attractors. By default, self loops are avoided.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>geneProbabilities</code></td>\n",
       "<td>\n",
       "\n",
       "<p>If <code>type=\"asynchronous\"</code>, this allows to specify probabilities for the genes. By default, each gene has the same probability to be chosen for the next state transition. You can supply a vector of probabilities for each of the genes which sums up to one.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>maxAttractorLength</code></td>\n",
       "<td>\n",
       "<p>If <code>method=\"sat.restricted\"</code>, this required parameter specifies the maximum size of attractors (i.e. the number of states in the loop) to be searched. For <code>method=\"sat.exhaustive\"</code>, this parameter is optional and specifies the maximum attractor length for the initial length-restricted search phase that is performed to speed up the subsequent exhaustive search. In this case, changing this value might bring performance benefits, but does not change the results.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>returnTable</code></td>\n",
       "<td>\n",
       "\n",
       "<p>Specifies whether a transition table is included in the returned <code>AttractorInfo</code> structure. If <code>type=\"asynchronous\"</code> or <code>method=\"sat\"</code>, this parameter is ignored, as the corresponding algorithms never return a transition table.\n",
       "</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>Depending on the type of network and the chosen parameters, different search algorithms are started.\n",
       "</p>\n",
       "<p>For <code>BooleanNetwork</code> networks, there are three different modes of attractor search:\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt>Exhaustive synchronous state space search</dt><dd>\n",
       "<p>In this mode, synchronous state transitions are carried out from each of the possible states until an attractor is reached. This identifies all synchronous attractors.\n",
       "</p>\n",
       "</dd>\n",
       "<dt>Heuristic synchronous state space search</dt><dd>\n",
       "<p>In contrast to exhaustive synchronous search, only a subset of the possible states is used. From these states, synchronous transitions are carried out until an attractor is reached. This subset is specified in <code>startStates</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt>Exhaustive synchronous SAT-based search</dt><dd>\n",
       "<p>Here, the attractor search problem is formulated as a satisfiability problem and solved using Armin Biere's PicoSAT solver. The algorithm is a variant of the method by Dubrova and Teslenko which searches for a satisfying assignment of a chain constructed by unfolding the transition relation. Depending on <code>maxAttractorLength</code>, it additionally applies an initial size-restricted SAT-based search (see below) to increase overall search speed. This method is suitable for larger networks of up to several hundreds of genes and exhaustively identifies all attractors in these networks. In contrast to the state space search, it does not construct and return a state transition table.\n",
       "</p>\n",
       "</dd>\n",
       "<dt>Size-restricted synchronous SAT-based search</dt><dd>\n",
       "<p>Here, the SAT solver directly looks for satisfying assignments for loops of a specific size. This may be more efficient for large networks and is guaranteed to find all attractors that comprise up to <code>maxAttractorLength</code> states (e.g. all steady states for <code>maxAttractorLength=1</code>) , but does not find any larger attractors. As for the exhaustive SAT-based method, no transition table is returned.\n",
       "</p>\n",
       "</dd>\n",
       "<dt>Heuristic asynchronous search</dt><dd>\n",
       "<p>This algorithm uses asynchronous state transitions and is able to identify steady-state and complex/loose attractors (see Harvey and Bossomaier, Garg et al.). These attractors are sets of states from which all possible asynchronous transitions lead into a state that is member of the set as well.\n",
       "The heuristic algorithm does the following for each of the input state specified by <code>startStates</code>:\n",
       "</p>\n",
       "\n",
       "<ol>\n",
       "<li><p> Perform <code>randomChainLength</code> random asynchronous transitions. After these transitions, the network state is expected to be located in an attractor with a high probability.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> Calculate the forward reachable set of the current state. Then, compare this set to the forward reachable set of all states in the set. If all sets are equal, a complex attractor is found.\n",
       "</p>\n",
       "</li></ol>\n",
       "\n",
       "</dd>\n",
       "</dl>\n",
       "<p>For <code>SymbolicBooleanNetwork</code> networks, <code>getAttractors</code> is simply a wrapper for <code>simulateSymbolicModel</code> with preset parameters. \n",
       "\n",
       "</p>\n",
       "<p>Printing the return value of <code>getAttractors</code> using <code>print</code> visualizes the identified attractors.</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>For <code>BooleanNetwork</code> networks, this returns a list of class <code>AttractorInfo</code> with components\n",
       "</p>\n",
       "<table summary=\"R valueblock\">\n",
       "<tr valign=\"top\"><td><code>attractors</code></td>\n",
       "<td>\n",
       "<p>A list of attractors. Each element is a 2-element list with the following components:\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt>involvedStates</dt><dd><p>A matrix containing the states that make up the attractor. Each column represents one state. The entries are decimal numbers that internally represent the states of the genes. The number of rows depends on the number of genes in the network: The first 32 genes are encoded in the first row, genes 33-64 are encoded in the second row, etc.</p>\n",
       "</dd>\n",
       "<dt>initialStates</dt><dd><p>This element is only available if an asynchronous search was carried out and this is a complex attractor. In this case, it holds the encoded start states of the transitions in the complex attractor</p>\n",
       "</dd>\n",
       "<dt>nextStates</dt><dd><p>This element is only available if an asynchronous search was carried out and this is a complex attractor. In this case, it holds the encoded successor states of the transitions in the complex attractor</p>\n",
       "</dd>\n",
       "<dt>basinSize</dt><dd><p>The number of states in the basin of attraction. Details on the states in the basin can be retrieved via <code>getBasinOfAttraction</code>.</p>\n",
       "</dd>\n",
       "</dl>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>stateInfo</code></td>\n",
       "<td>\n",
       "<p>A summary structure of class <code>BooleanStateInfo</code> containing information on the transition table. It has the following components:\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt>initialStates</dt><dd><p>This element is only available if <code>type=\"synchronous\"</code>, <code>method</code> is &quot;random&quot; or &quot;chosen&quot;, and <code>returnTable=TRUE</code>. This is a matrix describing the initial states that lead to the states in <code>table</code> after a state transition.\tIf <code>method</code> is &quot;exhaustive&quot;, this component is <code>NULL</code>. In this case, the initial states can be inferred, as all states are used. The format of the matrix is described in <code>involvedStates</code>.</p>\n",
       "</dd>\n",
       "<dt>table</dt><dd><p>This element is only available if <code>type=\"synchronous\"</code> and <code>returnTable=TRUE</code>. It holds result vector of the transition table as a matrix with one column for each state. These are encoded bit vectors in decimal numbers as described above.</p>\n",
       "</dd>\n",
       "<dt>attractorAssignment</dt><dd><p>This element is only available if <code>type=\"synchronous\"</code> and <code>returnTable=TRUE</code>. It contains a vector that corresponds to the entries in <code>table</code> and describes the attractor index in <code>attractors</code> to which successive transitions from the described state finally lead.</p>\n",
       "</dd>\n",
       "<dt>stepsToAttractor</dt><dd><p>This element is only available if <code>type=\"synchronous\"</code> and <code>returnTable=TRUE</code>. Referring to <code>attractorAssignment</code>, this is the number of transitions needed to reach the attractor.</p>\n",
       "</dd>\n",
       "<dt>genes</dt><dd><p>A list of names of the genes in <code>network</code>.</p>\n",
       "</dd>\n",
       "<dt>fixedGenes</dt><dd><p>Specifies the fixed genes as in the <code>fixed</code> component of <code>network</code>.</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "<p>The structure supports pretty printing using the <code>print</code> method.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "<p>For <code>SymbolicBooleanNetwork</code> networks, <code>getAttractors</code> redirects the call to <code>simulateSymbolicModel</code> and returns an object of class <code>SymbolicSimulation</code> containing the attractors and (if <code>returnTable=TRUE</code>) the transition graph.  \n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>S. A. Kauffman (1969), Metabolic stability and epigenesis in randomly constructed nets. J. Theor. Biol. 22:437&ndash;467.\n",
       "</p>\n",
       "<p>S. A. Kauffman (1993), The Origins of Order. Oxford University Press.\n",
       "</p>\n",
       "<p>I. Harvey, T. Bossomaier (1997), Time out of joint: Attractors in asynchronous random Boolean networks. Proc. of the Fourth European Conference on\n",
       "Artificial Life, 67&ndash;75.\n",
       "</p>\n",
       "<p>A. Garg, A. Di Cara, I. Xenarios, L. Mendoza, G. De Micheli (2008), Synchronous versus asynchronous modeling of gene regulatory networks. Bioinformatics 24(17):1917&ndash;1925. \n",
       "</p>\n",
       "<p>E. Dubrova, M. Teslenko (2011), A SAT-based algorithm for finding attractors in synchronous Boolean networks. IEEE/ACM Transactions on Computational Biology and Bioinformatics 8(5):1393&ndash;1399.\n",
       "</p>\n",
       "<p>A. Biere (2008), PicoSAT Essentials. Journal on Satisfiability, Boolean Modeling and Computation 4:75-97.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>loadNetwork</code>, <code>generateRandomNKNetwork</code>, <code>simulateSymbolicModel</code>, <code>plotAttractors</code>, <code>attractorsToLaTeX</code>, <code>getTransitionTable</code>, <code>getBasinOfAttraction</code>, <code>getAttractorSequence</code>, <code>getStateSummary</code>, <code>getPathToAttractor</code>, <code>fixGenes</code>, <code>generateState</code>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "# load example data\n",
       "data(cellcycle)\n",
       "\n",
       "# get all synchronous attractors by exhaustive search\n",
       "attractors &lt;- getAttractors(cellcycle)\n",
       "\n",
       "# plot attractors side by side\n",
       "par(mfrow=c(2, length(attractors$attractors)))\n",
       "plotAttractors(attractors)\n",
       "\n",
       "# finds the synchronous attractor with 7 states\n",
       "attractors &lt;- getAttractors(cellcycle, method=\"chosen\",\n",
       "                            startStates=list(rep(1, length(cellcycle$genes))))\n",
       "plotAttractors(attractors)\n",
       "\n",
       "# finds the attractor with 1 state\n",
       "attractors &lt;- getAttractors(cellcycle, method=\"chosen\",\n",
       "                            startStates=list(rep(0, length(cellcycle$genes))))\n",
       "plotAttractors(attractors)\n",
       "\n",
       "# also finds the attractor with 1 state by restricting the attractor length\n",
       "attractors &lt;- getAttractors(cellcycle, method=\"sat.restricted\",\n",
       "                            maxAttractorLength=1)\n",
       "plotAttractors(attractors)\n",
       "\n",
       "# identifies asynchronous attractors\n",
       "attractors &lt;- getAttractors(cellcycle, type=\"asynchronous\", startStates=100)\n",
       "plotAttractors(attractors, mode=\"graph\")\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>BoolNet</em> version 2.1.1 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{getAttractors}{Identify attractors in a Boolean network}{getAttractors}\n",
       "\\keyword{Boolean network}{getAttractors}\n",
       "\\keyword{synchronous update}{getAttractors}\n",
       "\\keyword{asynchronous update}{getAttractors}\n",
       "\\keyword{symbolic Boolean network}{getAttractors}\n",
       "\\keyword{attractor}{getAttractors}\n",
       "\\keyword{cycle}{getAttractors}\n",
       "\\keyword{basin}{getAttractors}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Identifies attractors (cycles) in a supplied Boolean network using synchronous or asynchronous state transitions\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "getAttractors(network, \n",
       "              type = c(\"synchronous\",\"asynchronous\"), \n",
       "              method = c(\"exhaustive\",\n",
       "                         \"sat.exhaustive\",\n",
       "                         \"sat.restricted\",\n",
       "                         \"random\",\n",
       "                         \"chosen\"), \n",
       "              startStates = list(),\n",
       "              genesON = c(), genesOFF = c(), \n",
       "              canonical = TRUE,\n",
       "              randomChainLength = 10000, \n",
       "              avoidSelfLoops = TRUE, \n",
       "              geneProbabilities = NULL, \n",
       "              maxAttractorLength = Inf,\n",
       "              returnTable = TRUE) \n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{network}] \n",
       "A network structure of class \\code{BooleanNetwork} or \\code{SymbolicBooleanNetwork}. These networks can be read from files by \\code{\\LinkA{loadNetwork}{loadNetwork}}, generated by \\code{\\LinkA{generateRandomNKNetwork}{generateRandomNKNetwork}}, or reconstructed by \\code{\\LinkA{reconstructNetwork}{reconstructNetwork}}.\n",
       "\n",
       "\n",
       "\\item[\\code{type}] \n",
       "If \\code{type=\"synchronous\"}, synchronous state transitions are used, i.e. all genes are updated at the same time. Synchronous attractor search can be performed in an exhaustive manner or using a heuristic that starts from predefined states. For symbolic networks, only synchronous updates are possible.\n",
       "\n",
       "If \\code{type=\"asynchronous\"}, asynchronous state transitions are performed, i.e. one (randomly chosen) gene is updated in each transition. Steady-state attractors are the same in asynchronous and synchronous networks, but the asynchronous search is also able to identify complex/loose attractors. Asynchronous search relies on a heuristic algorithm that starts from predefined states.\n",
       "\n",
       "See Details for more information on the algorithms.\n",
       "\n",
       "\n",
       "\n",
       "\\item[\\code{method}] \n",
       "The search method to be used. If \"exhaustive\", attractors are identified by exhaustive state space search, i.e. by calculating the sucessors of all 2\\textasciicircum{}n states (where n is the number of genes that are not set to a fixed value). This kind of search is only available for synchronous attractor search, and the maximum number of genes allowed for exhaustive search is 29. Apart from the attractors, this method generates the full state transition graph.\n",
       "\n",
       "If \\code{method} is \"sat.exhaustive\" or \"sat.restricted\", attractors are identified using algorithms based on the satisfiability problem. This search type is also restricted to synchronous networks. It can be used to identify attractors in much larger networks than with \\code{method=\"exhaustive\"}, but does not return the state transition graph. For \\code{method=\"sat.exhaustive\"}, an exhaustive attractor search is performed, while \\code{method=\"sat.restricted\"} only searches for attractors of a specified maximum length \\code{maxAttractorLength}.\n",
       "\n",
       "If \\code{method} is \"random\", \\code{startStates} is interpreted as an integer value specifying the number of states to be generated randomly. The algorithm is then initialized with these random states and identifies the attractors to which these states lead.\n",
       "\n",
       "If \\code{method} is \"chosen\", \\code{startStates} is interpreted as a list of binary vectors, each specifying one input state. Each vector must have \\code{length(network\\$genes)} elements with 0 or 1 values. The algorithm identifies the attractors to which the supplied states lead. If \\code{network} is of class \\code{SymbolicBooleanNetwork} and makes use of more than one predecessor state, this can also be a list of matrices with the genes in the columns and multiple predecessor states in the rows.\n",
       "\n",
       "If \\code{method} is not supplied, the desired method is inferred from the type of \\code{startStates}. By default, if neither \\code{method} nor \\code{startStates} are provided, an exhaustive search is performed.\n",
       "\n",
       "\n",
       "\\item[\\code{startStates}] \n",
       "The value of \\code{startStates} depends on the chosen method. See \\code{method} for more details.\n",
       " \n",
       "\\item[\\code{genesON}] \n",
       "A vector of genes whose values are fixed to 1, which reduces the complexity of the search. This is equivalent to a preceding call of \\code{\\LinkA{fixGenes}{fixGenes}}.\n",
       "\n",
       "\\item[\\code{genesOFF}] \n",
       "A vector of genes whose values are fixed to 0, which reduces the complexity of the search. This is equivalent to a preceding call of \\code{\\LinkA{fixGenes}{fixGenes}}.\n",
       "\n",
       "\\item[\\code{canonical}] \n",
       "If set to true, the states in the attractors are rearranged such that the state whose binary encoding\n",
       "makes up the smallest number is the first element of the vector. This ensures that attractors found by different heuristic runs of \\code{getAttractors} are comparable, as the cycles may have been entered at different states in different runs of the algorithm.\n",
       "\n",
       "\n",
       "\\item[\\code{randomChainLength}] \n",
       "If \\code{type=\"asynchronous\"}, this parameter specifies the number of random transitions performed by the search to enter a potential attractor (see Details). Defaults to 10000.\n",
       "\n",
       "\n",
       "\\item[\\code{avoidSelfLoops}] \n",
       "If \\code{type=\"asynchronous\"} and \\code{avoidSelfLoops=TRUE}, the asynchronous attractor search only enters self loops (i.e. transitions that result in the same state) if none of the possible transitions can leave the state. This results in attractors with fewer edges. Otherwise, self loops are included in the attractors. By default, self loops are avoided.\n",
       "\n",
       "\\item[\\code{geneProbabilities}] \n",
       "If \\code{type=\"asynchronous\"}, this allows to specify probabilities for the genes. By default, each gene has the same probability to be chosen for the next state transition. You can supply a vector of probabilities for each of the genes which sums up to one.\n",
       "\n",
       "\n",
       "\\item[\\code{maxAttractorLength}] If \\code{method=\"sat.restricted\"}, this required parameter specifies the maximum size of attractors (i.e. the number of states in the loop) to be searched. For \\code{method=\"sat.exhaustive\"}, this parameter is optional and specifies the maximum attractor length for the initial length-restricted search phase that is performed to speed up the subsequent exhaustive search. In this case, changing this value might bring performance benefits, but does not change the results.\n",
       "\n",
       "\\item[\\code{returnTable}] \n",
       "Specifies whether a transition table is included in the returned \\code{AttractorInfo} structure. If \\code{type=\"asynchronous\"} or \\code{method=\"sat\"}, this parameter is ignored, as the corresponding algorithms never return a transition table.\n",
       "\n",
       "\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "Depending on the type of network and the chosen parameters, different search algorithms are started.\n",
       "\n",
       "For \\code{BooleanNetwork} networks, there are three different modes of attractor search:\n",
       "\\begin{description}\n",
       "\n",
       "\\item[Exhaustive synchronous state space search] \n",
       "In this mode, synchronous state transitions are carried out from each of the possible states until an attractor is reached. This identifies all synchronous attractors.\n",
       "\n",
       "\n",
       "\\item[Heuristic synchronous state space search] \n",
       "In contrast to exhaustive synchronous search, only a subset of the possible states is used. From these states, synchronous transitions are carried out until an attractor is reached. This subset is specified in \\code{startStates}.\n",
       "\n",
       "\n",
       "\\item[Exhaustive synchronous SAT-based search] \n",
       "Here, the attractor search problem is formulated as a satisfiability problem and solved using Armin Biere's PicoSAT solver. The algorithm is a variant of the method by Dubrova and Teslenko which searches for a satisfying assignment of a chain constructed by unfolding the transition relation. Depending on \\code{maxAttractorLength}, it additionally applies an initial size-restricted SAT-based search (see below) to increase overall search speed. This method is suitable for larger networks of up to several hundreds of genes and exhaustively identifies all attractors in these networks. In contrast to the state space search, it does not construct and return a state transition table.\n",
       "\n",
       "\n",
       "\\item[Size-restricted synchronous SAT-based search] \n",
       "Here, the SAT solver directly looks for satisfying assignments for loops of a specific size. This may be more efficient for large networks and is guaranteed to find all attractors that comprise up to \\code{maxAttractorLength} states (e.g. all steady states for \\code{maxAttractorLength=1}) , but does not find any larger attractors. As for the exhaustive SAT-based method, no transition table is returned.\n",
       "\n",
       "\n",
       "\\item[Heuristic asynchronous search] \n",
       "This algorithm uses asynchronous state transitions and is able to identify steady-state and complex/loose attractors (see Harvey and Bossomaier, Garg et al.). These attractors are sets of states from which all possible asynchronous transitions lead into a state that is member of the set as well.\n",
       "The heuristic algorithm does the following for each of the input state specified by \\code{startStates}:\n",
       "\n",
       "\\begin{enumerate}\n",
       "\n",
       "\\item Perform \\code{randomChainLength} random asynchronous transitions. After these transitions, the network state is expected to be located in an attractor with a high probability.\n",
       "\n",
       "\\item Calculate the forward reachable set of the current state. Then, compare this set to the forward reachable set of all states in the set. If all sets are equal, a complex attractor is found.\n",
       "\n",
       "\\end{enumerate}\n",
       "\n",
       "\n",
       "\n",
       "For \\code{SymbolicBooleanNetwork} networks, \\code{getAttractors} is simply a wrapper for \\code{\\LinkA{simulateSymbolicModel}{simulateSymbolicModel}} with preset parameters. \n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\n",
       "Printing the return value of \\code{getAttractors} using \\code{\\LinkA{print}{print}} visualizes the identified attractors.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "\n",
       "For \\code{BooleanNetwork} networks, this returns a list of class \\code{AttractorInfo} with components\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{attractors}] A list of attractors. Each element is a 2-element list with the following components:\n",
       "\\begin{description}\n",
       "\n",
       "\\item[involvedStates] A matrix containing the states that make up the attractor. Each column represents one state. The entries are decimal numbers that internally represent the states of the genes. The number of rows depends on the number of genes in the network: The first 32 genes are encoded in the first row, genes 33-64 are encoded in the second row, etc.\n",
       "\\item[initialStates] This element is only available if an asynchronous search was carried out and this is a complex attractor. In this case, it holds the encoded start states of the transitions in the complex attractor\n",
       "\\item[nextStates] This element is only available if an asynchronous search was carried out and this is a complex attractor. In this case, it holds the encoded successor states of the transitions in the complex attractor\n",
       "\\item[basinSize] The number of states in the basin of attraction. Details on the states in the basin can be retrieved via \\code{\\LinkA{getBasinOfAttraction}{getBasinOfAttraction}}.\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\item[\\code{stateInfo}] A summary structure of class \\code{BooleanStateInfo} containing information on the transition table. It has the following components:\n",
       "\\begin{description}\n",
       "\n",
       "\\item[initialStates] This element is only available if \\code{type=\"synchronous\"}, \\code{method} is \"random\" or \"chosen\", and \\code{returnTable=TRUE}. This is a matrix describing the initial states that lead to the states in \\code{table} after a state transition.\tIf \\code{method} is \"exhaustive\", this component is \\code{NULL}. In this case, the initial states can be inferred, as all states are used. The format of the matrix is described in \\code{involvedStates}.\n",
       "\\item[table] This element is only available if \\code{type=\"synchronous\"} and \\\\{}\\code{returnTable=TRUE}. It holds result vector of the transition table as a matrix with one column for each state. These are encoded bit vectors in decimal numbers as described above.\n",
       "\\item[attractorAssignment] This element is only available if \\code{type=\"synchronous\"} and \\code{returnTable=TRUE}. It contains a vector that corresponds to the entries in \\code{table} and describes the attractor index in \\code{attractors} to which successive transitions from the described state finally lead.\n",
       "\\item[stepsToAttractor] This element is only available if \\code{type=\"synchronous\"} and \\code{returnTable=TRUE}. Referring to \\code{attractorAssignment}, this is the number of transitions needed to reach the attractor.\n",
       "\\item[genes] A list of names of the genes in \\code{network}.\n",
       "\\item[fixedGenes] Specifies the fixed genes as in the \\code{fixed} component of \\code{network}.\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "The structure supports pretty printing using the \\code{\\LinkA{print}{print}} method.\n",
       "\n",
       "\\end{ldescription}\n",
       "For \\code{SymbolicBooleanNetwork} networks, \\code{getAttractors} redirects the call to \\code{\\LinkA{simulateSymbolicModel}{simulateSymbolicModel}} and returns an object of class \\code{SymbolicSimulation} containing the attractors and (if \\code{returnTable=TRUE}) the transition graph.  \n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "S. A. Kauffman (1969), Metabolic stability and epigenesis in randomly constructed nets. J. Theor. Biol. 22:437--467.\n",
       "\n",
       "S. A. Kauffman (1993), The Origins of Order. Oxford University Press.\n",
       "\n",
       "I. Harvey, T. Bossomaier (1997), Time out of joint: Attractors in asynchronous random Boolean networks. Proc. of the Fourth European Conference on\n",
       "Artificial Life, 67--75.\n",
       "\n",
       "A. Garg, A. Di Cara, I. Xenarios, L. Mendoza, G. De Micheli (2008), Synchronous versus asynchronous modeling of gene regulatory networks. Bioinformatics 24(17):1917--1925. \n",
       "\n",
       "E. Dubrova, M. Teslenko (2011), A SAT-based algorithm for finding attractors in synchronous Boolean networks. IEEE/ACM Transactions on Computational Biology and Bioinformatics 8(5):1393--1399.\n",
       "\n",
       "A. Biere (2008), PicoSAT Essentials. Journal on Satisfiability, Boolean Modeling and Computation 4:75-97.\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{\\LinkA{loadNetwork}{loadNetwork}}, \\code{\\LinkA{generateRandomNKNetwork}{generateRandomNKNetwork}}, \\code{\\LinkA{simulateSymbolicModel}{simulateSymbolicModel}}, \\code{\\LinkA{plotAttractors}{plotAttractors}}, \\code{\\LinkA{attractorsToLaTeX}{attractorsToLaTeX}}, \\code{\\LinkA{getTransitionTable}{getTransitionTable}}, \\code{\\LinkA{getBasinOfAttraction}{getBasinOfAttraction}}, \\code{\\LinkA{getAttractorSequence}{getAttractorSequence}}, \\code{\\LinkA{getStateSummary}{getStateSummary}}, \\code{\\LinkA{getPathToAttractor}{getPathToAttractor}}, \\code{\\LinkA{fixGenes}{fixGenes}}, \\code{\\LinkA{generateState}{generateState}}\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "# load example data\n",
       "data(cellcycle)\n",
       "\n",
       "# get all synchronous attractors by exhaustive search\n",
       "attractors <- getAttractors(cellcycle)\n",
       "\n",
       "# plot attractors side by side\n",
       "par(mfrow=c(2, length(attractors$attractors)))\n",
       "plotAttractors(attractors)\n",
       "\n",
       "# finds the synchronous attractor with 7 states\n",
       "attractors <- getAttractors(cellcycle, method=\"chosen\",\n",
       "                            startStates=list(rep(1, length(cellcycle$genes))))\n",
       "plotAttractors(attractors)\n",
       "\n",
       "# finds the attractor with 1 state\n",
       "attractors <- getAttractors(cellcycle, method=\"chosen\",\n",
       "                            startStates=list(rep(0, length(cellcycle$genes))))\n",
       "plotAttractors(attractors)\n",
       "\n",
       "# also finds the attractor with 1 state by restricting the attractor length\n",
       "attractors <- getAttractors(cellcycle, method=\"sat.restricted\",\n",
       "                            maxAttractorLength=1)\n",
       "plotAttractors(attractors)\n",
       "\n",
       "# identifies asynchronous attractors\n",
       "attractors <- getAttractors(cellcycle, type=\"asynchronous\", startStates=100)\n",
       "plotAttractors(attractors, mode=\"graph\")\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "getAttractors             package:BoolNet              R Documentation\n",
       "\n",
       "_\bI_\bd_\be_\bn_\bt_\bi_\bf_\by _\ba_\bt_\bt_\br_\ba_\bc_\bt_\bo_\br_\bs _\bi_\bn _\ba _\bB_\bo_\bo_\bl_\be_\ba_\bn _\bn_\be_\bt_\bw_\bo_\br_\bk\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Identifies attractors (cycles) in a supplied Boolean network using\n",
       "     synchronous or asynchronous state transitions\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     getAttractors(network, \n",
       "                   type = c(\"synchronous\",\"asynchronous\"), \n",
       "                   method = c(\"exhaustive\",\n",
       "                              \"sat.exhaustive\",\n",
       "                              \"sat.restricted\",\n",
       "                              \"random\",\n",
       "                              \"chosen\"), \n",
       "                   startStates = list(),\n",
       "                   genesON = c(), genesOFF = c(), \n",
       "                   canonical = TRUE,\n",
       "                   randomChainLength = 10000, \n",
       "                   avoidSelfLoops = TRUE, \n",
       "                   geneProbabilities = NULL, \n",
       "                   maxAttractorLength = Inf,\n",
       "                   returnTable = TRUE) \n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       " network: A network structure of class ‘BooleanNetwork’ or\n",
       "          ‘SymbolicBooleanNetwork’. These networks can be read from\n",
       "          files by ‘loadNetwork’, generated by\n",
       "          ‘generateRandomNKNetwork’, or reconstructed by\n",
       "          ‘reconstructNetwork’.\n",
       "\n",
       "    type: If ‘type=\"synchronous\"’, synchronous state transitions are\n",
       "          used, i.e. all genes are updated at the same time.\n",
       "          Synchronous attractor search can be performed in an\n",
       "          exhaustive manner or using a heuristic that starts from\n",
       "          predefined states. For symbolic networks, only synchronous\n",
       "          updates are possible.\n",
       "\n",
       "          If ‘type=\"asynchronous\"’, asynchronous state transitions are\n",
       "          performed, i.e. one (randomly chosen) gene is updated in each\n",
       "          transition. Steady-state attractors are the same in\n",
       "          asynchronous and synchronous networks, but the asynchronous\n",
       "          search is also able to identify complex/loose attractors.\n",
       "          Asynchronous search relies on a heuristic algorithm that\n",
       "          starts from predefined states.\n",
       "\n",
       "          See Details for more information on the algorithms.\n",
       "\n",
       "  method: The search method to be used. If \"exhaustive\", attractors are\n",
       "          identified by exhaustive state space search, i.e. by\n",
       "          calculating the sucessors of all 2^n states (where n is the\n",
       "          number of genes that are not set to a fixed value). This kind\n",
       "          of search is only available for synchronous attractor search,\n",
       "          and the maximum number of genes allowed for exhaustive search\n",
       "          is 29. Apart from the attractors, this method generates the\n",
       "          full state transition graph.\n",
       "\n",
       "          If ‘method’ is \"sat.exhaustive\" or \"sat.restricted\",\n",
       "          attractors are identified using algorithms based on the\n",
       "          satisfiability problem. This search type is also restricted\n",
       "          to synchronous networks. It can be used to identify\n",
       "          attractors in much larger networks than with\n",
       "          ‘method=\"exhaustive\"’, but does not return the state\n",
       "          transition graph. For ‘method=\"sat.exhaustive\"’, an\n",
       "          exhaustive attractor search is performed, while\n",
       "          ‘method=\"sat.restricted\"’ only searches for attractors of a\n",
       "          specified maximum length ‘maxAttractorLength’.\n",
       "\n",
       "          If ‘method’ is \"random\", ‘startStates’ is interpreted as an\n",
       "          integer value specifying the number of states to be generated\n",
       "          randomly. The algorithm is then initialized with these random\n",
       "          states and identifies the attractors to which these states\n",
       "          lead.\n",
       "\n",
       "          If ‘method’ is \"chosen\", ‘startStates’ is interpreted as a\n",
       "          list of binary vectors, each specifying one input state. Each\n",
       "          vector must have ‘length(network$genes)’ elements with 0 or 1\n",
       "          values. The algorithm identifies the attractors to which the\n",
       "          supplied states lead. If ‘network’ is of class\n",
       "          ‘SymbolicBooleanNetwork’ and makes use of more than one\n",
       "          predecessor state, this can also be a list of matrices with\n",
       "          the genes in the columns and multiple predecessor states in\n",
       "          the rows.\n",
       "\n",
       "          If ‘method’ is not supplied, the desired method is inferred\n",
       "          from the type of ‘startStates’. By default, if neither\n",
       "          ‘method’ nor ‘startStates’ are provided, an exhaustive search\n",
       "          is performed.\n",
       "\n",
       "startStates: The value of ‘startStates’ depends on the chosen method.\n",
       "          See ‘method’ for more details.\n",
       "\n",
       " genesON: A vector of genes whose values are fixed to 1, which reduces\n",
       "          the complexity of the search. This is equivalent to a\n",
       "          preceding call of ‘fixGenes’.\n",
       "\n",
       "genesOFF: A vector of genes whose values are fixed to 0, which reduces\n",
       "          the complexity of the search. This is equivalent to a\n",
       "          preceding call of ‘fixGenes’.\n",
       "\n",
       "canonical: If set to true, the states in the attractors are rearranged\n",
       "          such that the state whose binary encoding makes up the\n",
       "          smallest number is the first element of the vector. This\n",
       "          ensures that attractors found by different heuristic runs of\n",
       "          ‘getAttractors’ are comparable, as the cycles may have been\n",
       "          entered at different states in different runs of the\n",
       "          algorithm.\n",
       "\n",
       "randomChainLength: If ‘type=\"asynchronous\"’, this parameter specifies\n",
       "          the number of random transitions performed by the search to\n",
       "          enter a potential attractor (see Details). Defaults to 10000.\n",
       "\n",
       "avoidSelfLoops: If ‘type=\"asynchronous\"’ and ‘avoidSelfLoops=TRUE’, the\n",
       "          asynchronous attractor search only enters self loops (i.e.\n",
       "          transitions that result in the same state) if none of the\n",
       "          possible transitions can leave the state. This results in\n",
       "          attractors with fewer edges. Otherwise, self loops are\n",
       "          included in the attractors. By default, self loops are\n",
       "          avoided.\n",
       "\n",
       "geneProbabilities: If ‘type=\"asynchronous\"’, this allows to specify\n",
       "          probabilities for the genes. By default, each gene has the\n",
       "          same probability to be chosen for the next state transition.\n",
       "          You can supply a vector of probabilities for each of the\n",
       "          genes which sums up to one.\n",
       "\n",
       "maxAttractorLength: If ‘method=\"sat.restricted\"’, this required\n",
       "          parameter specifies the maximum size of attractors (i.e. the\n",
       "          number of states in the loop) to be searched. For\n",
       "          ‘method=\"sat.exhaustive\"’, this parameter is optional and\n",
       "          specifies the maximum attractor length for the initial\n",
       "          length-restricted search phase that is performed to speed up\n",
       "          the subsequent exhaustive search. In this case, changing this\n",
       "          value might bring performance benefits, but does not change\n",
       "          the results.\n",
       "\n",
       "returnTable: Specifies whether a transition table is included in the\n",
       "          returned ‘AttractorInfo’ structure. If ‘type=\"asynchronous\"’\n",
       "          or ‘method=\"sat\"’, this parameter is ignored, as the\n",
       "          corresponding algorithms never return a transition table.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     Depending on the type of network and the chosen parameters,\n",
       "     different search algorithms are started.\n",
       "\n",
       "     For ‘BooleanNetwork’ networks, there are three different modes of\n",
       "     attractor search:\n",
       "\n",
       "     Exhaustive synchronous state space search In this mode,\n",
       "          synchronous state transitions are carried out from each of\n",
       "          the possible states until an attractor is reached. This\n",
       "          identifies all synchronous attractors.\n",
       "\n",
       "     Heuristic synchronous state space search In contrast to exhaustive\n",
       "          synchronous search, only a subset of the possible states is\n",
       "          used. From these states, synchronous transitions are carried\n",
       "          out until an attractor is reached. This subset is specified\n",
       "          in ‘startStates’.\n",
       "\n",
       "     Exhaustive synchronous SAT-based search Here, the attractor search\n",
       "          problem is formulated as a satisfiability problem and solved\n",
       "          using Armin Biere's PicoSAT solver. The algorithm is a\n",
       "          variant of the method by Dubrova and Teslenko which searches\n",
       "          for a satisfying assignment of a chain constructed by\n",
       "          unfolding the transition relation. Depending on\n",
       "          ‘maxAttractorLength’, it additionally applies an initial\n",
       "          size-restricted SAT-based search (see below) to increase\n",
       "          overall search speed. This method is suitable for larger\n",
       "          networks of up to several hundreds of genes and exhaustively\n",
       "          identifies all attractors in these networks. In contrast to\n",
       "          the state space search, it does not construct and return a\n",
       "          state transition table.\n",
       "\n",
       "     Size-restricted synchronous SAT-based search Here, the SAT solver\n",
       "          directly looks for satisfying assignments for loops of a\n",
       "          specific size. This may be more efficient for large networks\n",
       "          and is guaranteed to find all attractors that comprise up to\n",
       "          ‘maxAttractorLength’ states (e.g. all steady states for\n",
       "          ‘maxAttractorLength=1’) , but does not find any larger\n",
       "          attractors. As for the exhaustive SAT-based method, no\n",
       "          transition table is returned.\n",
       "\n",
       "     Heuristic asynchronous search This algorithm uses asynchronous\n",
       "          state transitions and is able to identify steady-state and\n",
       "          complex/loose attractors (see Harvey and Bossomaier, Garg et\n",
       "          al.). These attractors are sets of states from which all\n",
       "          possible asynchronous transitions lead into a state that is\n",
       "          member of the set as well. The heuristic algorithm does the\n",
       "          following for each of the input state specified by\n",
       "          ‘startStates’:\n",
       "\n",
       "           1. Perform ‘randomChainLength’ random asynchronous\n",
       "              transitions. After these transitions, the network state\n",
       "              is expected to be located in an attractor with a high\n",
       "              probability.\n",
       "\n",
       "           2. Calculate the forward reachable set of the current state.\n",
       "              Then, compare this set to the forward reachable set of\n",
       "              all states in the set. If all sets are equal, a complex\n",
       "              attractor is found.\n",
       "\n",
       "     For ‘SymbolicBooleanNetwork’ networks, ‘getAttractors’ is simply a\n",
       "     wrapper for ‘simulateSymbolicModel’ with preset parameters.\n",
       "\n",
       "     Printing the return value of ‘getAttractors’ using ‘print’\n",
       "     visualizes the identified attractors.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     For ‘BooleanNetwork’ networks, this returns a list of class\n",
       "     ‘AttractorInfo’ with components\n",
       "\n",
       "attractors: A list of attractors. Each element is a 2-element list with\n",
       "          the following components:\n",
       "\n",
       "          involvedStates A matrix containing the states that make up\n",
       "              the attractor. Each column represents one state. The\n",
       "              entries are decimal numbers that internally represent the\n",
       "              states of the genes. The number of rows depends on the\n",
       "              number of genes in the network: The first 32 genes are\n",
       "              encoded in the first row, genes 33-64 are encoded in the\n",
       "              second row, etc.\n",
       "\n",
       "          initialStates This element is only available if an\n",
       "              asynchronous search was carried out and this is a complex\n",
       "              attractor. In this case, it holds the encoded start\n",
       "              states of the transitions in the complex attractor\n",
       "\n",
       "          nextStates This element is only available if an asynchronous\n",
       "              search was carried out and this is a complex attractor.\n",
       "              In this case, it holds the encoded successor states of\n",
       "              the transitions in the complex attractor\n",
       "\n",
       "          basinSize The number of states in the basin of attraction.\n",
       "              Details on the states in the basin can be retrieved via\n",
       "              ‘getBasinOfAttraction’.\n",
       "\n",
       "stateInfo: A summary structure of class ‘BooleanStateInfo’ containing\n",
       "          information on the transition table. It has the following\n",
       "          components:\n",
       "\n",
       "          initialStates This element is only available if\n",
       "              ‘type=\"synchronous\"’, ‘method’ is \"random\" or \"chosen\",\n",
       "              and ‘returnTable=TRUE’. This is a matrix describing the\n",
       "              initial states that lead to the states in ‘table’ after a\n",
       "              state transition.  If ‘method’ is \"exhaustive\", this\n",
       "              component is ‘NULL’. In this case, the initial states can\n",
       "              be inferred, as all states are used. The format of the\n",
       "              matrix is described in ‘involvedStates’.\n",
       "\n",
       "          table This element is only available if ‘type=\"synchronous\"’\n",
       "              and ‘returnTable=TRUE’. It holds result vector of the\n",
       "              transition table as a matrix with one column for each\n",
       "              state. These are encoded bit vectors in decimal numbers\n",
       "              as described above.\n",
       "\n",
       "          attractorAssignment This element is only available if\n",
       "              ‘type=\"synchronous\"’ and ‘returnTable=TRUE’. It contains\n",
       "              a vector that corresponds to the entries in ‘table’ and\n",
       "              describes the attractor index in ‘attractors’ to which\n",
       "              successive transitions from the described state finally\n",
       "              lead.\n",
       "\n",
       "          stepsToAttractor This element is only available if\n",
       "              ‘type=\"synchronous\"’ and ‘returnTable=TRUE’. Referring to\n",
       "              ‘attractorAssignment’, this is the number of transitions\n",
       "              needed to reach the attractor.\n",
       "\n",
       "          genes A list of names of the genes in ‘network’.\n",
       "\n",
       "          fixedGenes Specifies the fixed genes as in the ‘fixed’\n",
       "              component of ‘network’.\n",
       "\n",
       "          The structure supports pretty printing using the ‘print’\n",
       "          method.\n",
       "     For ‘SymbolicBooleanNetwork’ networks, ‘getAttractors’ redirects\n",
       "     the call to ‘simulateSymbolicModel’ and returns an object of class\n",
       "     ‘SymbolicSimulation’ containing the attractors and (if\n",
       "     ‘returnTable=TRUE’) the transition graph.\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     S. A. Kauffman (1969), Metabolic stability and epigenesis in\n",
       "     randomly constructed nets. J. Theor. Biol. 22:437-467.\n",
       "\n",
       "     S. A. Kauffman (1993), The Origins of Order. Oxford University\n",
       "     Press.\n",
       "\n",
       "     I. Harvey, T. Bossomaier (1997), Time out of joint: Attractors in\n",
       "     asynchronous random Boolean networks. Proc. of the Fourth European\n",
       "     Conference on Artificial Life, 67-75.\n",
       "\n",
       "     A. Garg, A. Di Cara, I. Xenarios, L. Mendoza, G. De Micheli\n",
       "     (2008), Synchronous versus asynchronous modeling of gene\n",
       "     regulatory networks. Bioinformatics 24(17):1917-1925.\n",
       "\n",
       "     E. Dubrova, M. Teslenko (2011), A SAT-based algorithm for finding\n",
       "     attractors in synchronous Boolean networks. IEEE/ACM Transactions\n",
       "     on Computational Biology and Bioinformatics 8(5):1393-1399.\n",
       "\n",
       "     A. Biere (2008), PicoSAT Essentials. Journal on Satisfiability,\n",
       "     Boolean Modeling and Computation 4:75-97.\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘loadNetwork’, ‘generateRandomNKNetwork’, ‘simulateSymbolicModel’,\n",
       "     ‘plotAttractors’, ‘attractorsToLaTeX’, ‘getTransitionTable’,\n",
       "     ‘getBasinOfAttraction’, ‘getAttractorSequence’, ‘getStateSummary’,\n",
       "     ‘getPathToAttractor’, ‘fixGenes’, ‘generateState’\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     # load example data\n",
       "     data(cellcycle)\n",
       "     \n",
       "     # get all synchronous attractors by exhaustive search\n",
       "     attractors <- getAttractors(cellcycle)\n",
       "     \n",
       "     # plot attractors side by side\n",
       "     par(mfrow=c(2, length(attractors$attractors)))\n",
       "     plotAttractors(attractors)\n",
       "     \n",
       "     # finds the synchronous attractor with 7 states\n",
       "     attractors <- getAttractors(cellcycle, method=\"chosen\",\n",
       "                                 startStates=list(rep(1, length(cellcycle$genes))))\n",
       "     plotAttractors(attractors)\n",
       "     \n",
       "     # finds the attractor with 1 state\n",
       "     attractors <- getAttractors(cellcycle, method=\"chosen\",\n",
       "                                 startStates=list(rep(0, length(cellcycle$genes))))\n",
       "     plotAttractors(attractors)\n",
       "     \n",
       "     # also finds the attractor with 1 state by restricting the attractor length\n",
       "     attractors <- getAttractors(cellcycle, method=\"sat.restricted\",\n",
       "                                 maxAttractorLength=1)\n",
       "     plotAttractors(attractors)\n",
       "     \n",
       "     # identifies asynchronous attractors\n",
       "     attractors <- getAttractors(cellcycle, type=\"asynchronous\", startStates=100)\n",
       "     plotAttractors(attractors, mode=\"graph\")\n",
       "     "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "?getAttractors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we will use the default options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Attractor 1 is a simple attractor consisting of 1 state(s) and has a basin of 27 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000000000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 2 is a simple attractor consisting of 1 state(s) and has a basin of 2 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000000000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 3 is a simple attractor consisting of 1 state(s) and has a basin of 14 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0010000000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 4 is a simple attractor consisting of 1 state(s) and has a basin of 13 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000100000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 5 is a simple attractor consisting of 1 state(s) and has a basin of 8 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000110000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 6 is a simple attractor consisting of 1 state(s) and has a basin of 35 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000001000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 7 is a simple attractor consisting of 1 state(s) and has a basin of 8 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0010001000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 8 is a simple attractor consisting of 1 state(s) and has a basin of 13 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001101000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 9 is a simple attractor consisting of 1 state(s) and has a basin of 3 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000111000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 10 is a simple attractor consisting of 1 state(s) and has a basin of 1 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001111000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 11 is a simple attractor consisting of 1 state(s) and has a basin of 54 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0010000100   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 12 is a simple attractor consisting of 1 state(s) and has a basin of 2 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000000100   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 13 is a simple attractor consisting of 1 state(s) and has a basin of 8 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000110100   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 14 is a simple attractor consisting of 1 state(s) and has a basin of 2 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000001100   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 15 is a simple attractor consisting of 1 state(s) and has a basin of 8 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0010001100   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 16 is a simple attractor consisting of 1 state(s) and has a basin of 10 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001101100   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 17 is a simple attractor consisting of 1 state(s) and has a basin of 3 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000111100   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 18 is a simple attractor consisting of 1 state(s) and has a basin of 1 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001111100   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 19 is a simple attractor consisting of 1 state(s) and has a basin of 26 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000100010   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 20 is a simple attractor consisting of 1 state(s) and has a basin of 14 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0110100010   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 21 is a simple attractor consisting of 1 state(s) and has a basin of 24 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000110010   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 22 is a simple attractor consisting of 1 state(s) and has a basin of 26 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001101010   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 23 is a simple attractor consisting of 1 state(s) and has a basin of 8 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0110101010   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 24 is a simple attractor consisting of 1 state(s) and has a basin of 19 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000111010   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 25 is a simple attractor consisting of 1 state(s) and has a basin of 1 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001111010   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 26 is a simple attractor consisting of 1 state(s) and has a basin of 40 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0110100110   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 27 is a simple attractor consisting of 1 state(s) and has a basin of 24 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000110110   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 28 is a simple attractor consisting of 1 state(s) and has a basin of 12 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001101110   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 29 is a simple attractor consisting of 1 state(s) and has a basin of 8 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0110101110   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 30 is a simple attractor consisting of 1 state(s) and has a basin of 19 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000111110   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 31 is a simple attractor consisting of 1 state(s) and has a basin of 1 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001111110   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 32 is a simple attractor consisting of 1 state(s) and has a basin of 43 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000010001   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 33 is a simple attractor consisting of 1 state(s) and has a basin of 21 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000110001   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 34 is a simple attractor consisting of 1 state(s) and has a basin of 43 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000011001   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 35 is a simple attractor consisting of 1 state(s) and has a basin of 5 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001111001   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 36 is a simple attractor consisting of 1 state(s) and has a basin of 3 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000111001   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 37 is a simple attractor consisting of 1 state(s) and has a basin of 46 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000010101   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 38 is a simple attractor consisting of 1 state(s) and has a basin of 18 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000110101   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 39 is a simple attractor consisting of 1 state(s) and has a basin of 46 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000011101   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 40 is a simple attractor consisting of 1 state(s) and has a basin of 5 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001111101   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 41 is a simple attractor consisting of 1 state(s) and has a basin of 3 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000111101   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 42 is a simple attractor consisting of 1 state(s) and has a basin of 64 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000110011   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 43 is a simple attractor consisting of 1 state(s) and has a basin of 43 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000111011   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 44 is a simple attractor consisting of 1 state(s) and has a basin of 5 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001111011   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 45 is a simple attractor consisting of 1 state(s) and has a basin of 64 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000110111   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 46 is a simple attractor consisting of 1 state(s) and has a basin of 47 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000111111   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 47 is a simple attractor consisting of 1 state(s) and has a basin of 5 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1001111111   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 48 is a simple attractor consisting of 2 state(s) and has a basin of 4 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000111000   |\n",
       " 0001111000   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 49 is a simple attractor consisting of 2 state(s) and has a basin of 36 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0000001100   |\n",
       " 1010001100   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 50 is a simple attractor consisting of 2 state(s) and has a basin of 4 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000111100   |\n",
       " 0001111100   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 51 is a simple attractor consisting of 2 state(s) and has a basin of 6 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0100101010   |\n",
       " 1010101010   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 52 is a simple attractor consisting of 2 state(s) and has a basin of 4 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000111010   |\n",
       " 0001111010   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 53 is a simple attractor consisting of 2 state(s) and has a basin of 20 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 0100101110   |\n",
       " 1010101110   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 54 is a simple attractor consisting of 2 state(s) and has a basin of 4 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000111110   |\n",
       " 0001111110   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 55 is a simple attractor consisting of 2 state(s) and has a basin of 13 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000111001   |\n",
       " 0001111001   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 56 is a simple attractor consisting of 2 state(s) and has a basin of 10 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000111101   |\n",
       " 0001111101   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 57 is a simple attractor consisting of 2 state(s) and has a basin of 16 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000111011   |\n",
       " 0001111011   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n",
       "\n",
       "Attractor 58 is a simple attractor consisting of 2 state(s) and has a basin of 12 state(s):\n",
       "\n",
       " |--<---------|\n",
       " V            |\n",
       " 1000111111   |\n",
       " 0001111111   |\n",
       " V            |\n",
       " |-->---------|\n",
       "\n",
       "\n",
       "Genes are encoded in the following order: IL2 RORGT STAT3 FOXP3 TGFB IL10 IL2e IL21e TGFBe IL10e\n"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attr <- getAttractors(net)\n",
    "attr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The result is an attractor object. States are saved as an integer. This object structure can change depending of the options used. The general structure is:\n",
    "* __attractor__\n",
    "    * __\\$stateInfo__ : information of the simulation, evaluated states and transition table\n",
    "        * __\\$table__,  __\\$attractorAssignment__,  __\\$stepsToAttractor__,  __\\$initialStates__,  __\\$genes__,  __\\$fixedGenes__,  \n",
    "    * __\\attractors$__ : list of attractors. (Use iterator or [[n]] to examine each node)\n",
    "        * __\\$involvedStates__ : states that make the attractor\n",
    "        * __\\$basinSize__ : states that reach the attractor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "$involvedStates\n",
       "     [,1] [,2]\n",
       "[1,] 1009 1016\n",
       "\n",
       "$basinSize\n",
       "[1] 12\n"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attr$attractors[[58]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can create a dataframe including only the state and the basin size. However, there are some attractors that have more than one state, which can complicate the process. To solve this we will create a function __attractor2dataframe()__ that represent the states as characters and join them with \"\\\" and returns a dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "attractor2dataframe <- function(attr) {\n",
    "    data.frame(\n",
    "        involvedStates = sapply(attr$attractors, function(a) {\n",
    "            paste(as.character(a$involvedStates), collapse='/')\n",
    "        })  , \n",
    "        basinSize = sapply(attr$attractors, function(a) a$basinSize )\n",
    ")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>involvedStates</th><th scope=col>basinSize</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>0</td><td>27</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>4</td><td>14</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>16</td><td>13</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>48</td><td>8</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>65</td><td>35</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>68</td><td>8</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>89</td><td>13</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>112</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>121</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>132</td><td>54</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>129</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>176</td><td>8</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>193</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>196</td><td>8</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>217</td><td>10</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>240</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>249</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>272</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>278</td><td>14</td></tr>\n",
       "\t<tr><th scope=row>21</th><td>304</td><td>24</td></tr>\n",
       "\t<tr><th scope=row>22</th><td>345</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>23</th><td>342</td><td>8</td></tr>\n",
       "\t<tr><th scope=row>24</th><td>368</td><td>19</td></tr>\n",
       "\t<tr><th scope=row>25</th><td>377</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>26</th><td>406</td><td>40</td></tr>\n",
       "\t<tr><th scope=row>27</th><td>432</td><td>24</td></tr>\n",
       "\t<tr><th scope=row>28</th><td>473</td><td>12</td></tr>\n",
       "\t<tr><th scope=row>29</th><td>470</td><td>8</td></tr>\n",
       "\t<tr><th scope=row>30</th><td>496</td><td>19</td></tr>\n",
       "\t<tr><th scope=row>31</th><td>505</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>32</th><td>544</td><td>43</td></tr>\n",
       "\t<tr><th scope=row>33</th><td>560</td><td>21</td></tr>\n",
       "\t<tr><th scope=row>34</th><td>608</td><td>43</td></tr>\n",
       "\t<tr><th scope=row>35</th><td>633</td><td>5</td></tr>\n",
       "\t<tr><th scope=row>36</th><td>624</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>37</th><td>672</td><td>46</td></tr>\n",
       "\t<tr><th scope=row>38</th><td>688</td><td>18</td></tr>\n",
       "\t<tr><th scope=row>39</th><td>736</td><td>46</td></tr>\n",
       "\t<tr><th scope=row>40</th><td>761</td><td>5</td></tr>\n",
       "\t<tr><th scope=row>41</th><td>752</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>42</th><td>816</td><td>64</td></tr>\n",
       "\t<tr><th scope=row>43</th><td>880</td><td>43</td></tr>\n",
       "\t<tr><th scope=row>44</th><td>889</td><td>5</td></tr>\n",
       "\t<tr><th scope=row>45</th><td>944</td><td>64</td></tr>\n",
       "\t<tr><th scope=row>46</th><td>1008</td><td>47</td></tr>\n",
       "\t<tr><th scope=row>47</th><td>1017</td><td>5</td></tr>\n",
       "\t<tr><th scope=row>48</th><td>113/120</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>49</th><td>192/197</td><td>36</td></tr>\n",
       "\t<tr><th scope=row>50</th><td>241/248</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>51</th><td>338/341</td><td>6</td></tr>\n",
       "\t<tr><th scope=row>52</th><td>369/376</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>53</th><td>466/469</td><td>20</td></tr>\n",
       "\t<tr><th scope=row>54</th><td>497/504</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>55</th><td>625/632</td><td>13</td></tr>\n",
       "\t<tr><th scope=row>56</th><td>753/760</td><td>10</td></tr>\n",
       "\t<tr><th scope=row>57</th><td>881/888</td><td>16</td></tr>\n",
       "\t<tr><th scope=row>58</th><td>1009/1016</td><td>12</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       "  & involvedStates & basinSize\\\\\n",
       "\\hline\n",
       "\t1 & 0 & 27\\\\\n",
       "\t2 & 1 & 2\\\\\n",
       "\t3 & 4 & 14\\\\\n",
       "\t4 & 16 & 13\\\\\n",
       "\t5 & 48 & 8\\\\\n",
       "\t6 & 65 & 35\\\\\n",
       "\t7 & 68 & 8\\\\\n",
       "\t8 & 89 & 13\\\\\n",
       "\t9 & 112 & 3\\\\\n",
       "\t10 & 121 & 1\\\\\n",
       "\t11 & 132 & 54\\\\\n",
       "\t12 & 129 & 2\\\\\n",
       "\t13 & 176 & 8\\\\\n",
       "\t14 & 193 & 2\\\\\n",
       "\t15 & 196 & 8\\\\\n",
       "\t16 & 217 & 10\\\\\n",
       "\t17 & 240 & 3\\\\\n",
       "\t18 & 249 & 1\\\\\n",
       "\t19 & 272 & 26\\\\\n",
       "\t20 & 278 & 14\\\\\n",
       "\t21 & 304 & 24\\\\\n",
       "\t22 & 345 & 26\\\\\n",
       "\t23 & 342 & 8\\\\\n",
       "\t24 & 368 & 19\\\\\n",
       "\t25 & 377 & 1\\\\\n",
       "\t26 & 406 & 40\\\\\n",
       "\t27 & 432 & 24\\\\\n",
       "\t28 & 473 & 12\\\\\n",
       "\t29 & 470 & 8\\\\\n",
       "\t30 & 496 & 19\\\\\n",
       "\t31 & 505 & 1\\\\\n",
       "\t32 & 544 & 43\\\\\n",
       "\t33 & 560 & 21\\\\\n",
       "\t34 & 608 & 43\\\\\n",
       "\t35 & 633 & 5\\\\\n",
       "\t36 & 624 & 3\\\\\n",
       "\t37 & 672 & 46\\\\\n",
       "\t38 & 688 & 18\\\\\n",
       "\t39 & 736 & 46\\\\\n",
       "\t40 & 761 & 5\\\\\n",
       "\t41 & 752 & 3\\\\\n",
       "\t42 & 816 & 64\\\\\n",
       "\t43 & 880 & 43\\\\\n",
       "\t44 & 889 & 5\\\\\n",
       "\t45 & 944 & 64\\\\\n",
       "\t46 & 1008 & 47\\\\\n",
       "\t47 & 1017 & 5\\\\\n",
       "\t48 & 113/120 & 4\\\\\n",
       "\t49 & 192/197 & 36\\\\\n",
       "\t50 & 241/248 & 4\\\\\n",
       "\t51 & 338/341 & 6\\\\\n",
       "\t52 & 369/376 & 4\\\\\n",
       "\t53 & 466/469 & 20\\\\\n",
       "\t54 & 497/504 & 4\\\\\n",
       "\t55 & 625/632 & 13\\\\\n",
       "\t56 & 753/760 & 10\\\\\n",
       "\t57 & 881/888 & 16\\\\\n",
       "\t58 & 1009/1016 & 12\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "   involvedStates basinSize\n",
       "1               0        27\n",
       "2               1         2\n",
       "3               4        14\n",
       "4              16        13\n",
       "5              48         8\n",
       "6              65        35\n",
       "7              68         8\n",
       "8              89        13\n",
       "9             112         3\n",
       "10            121         1\n",
       "11            132        54\n",
       "12            129         2\n",
       "13            176         8\n",
       "14            193         2\n",
       "15            196         8\n",
       "16            217        10\n",
       "17            240         3\n",
       "18            249         1\n",
       "19            272        26\n",
       "20            278        14\n",
       "21            304        24\n",
       "22            345        26\n",
       "23            342         8\n",
       "24            368        19\n",
       "25            377         1\n",
       "26            406        40\n",
       "27            432        24\n",
       "28            473        12\n",
       "29            470         8\n",
       "30            496        19\n",
       "31            505         1\n",
       "32            544        43\n",
       "33            560        21\n",
       "34            608        43\n",
       "35            633         5\n",
       "36            624         3\n",
       "37            672        46\n",
       "38            688        18\n",
       "39            736        46\n",
       "40            761         5\n",
       "41            752         3\n",
       "42            816        64\n",
       "43            880        43\n",
       "44            889         5\n",
       "45            944        64\n",
       "46           1008        47\n",
       "47           1017         5\n",
       "48        113/120         4\n",
       "49        192/197        36\n",
       "50        241/248         4\n",
       "51        338/341         6\n",
       "52        369/376         4\n",
       "53        466/469        20\n",
       "54        497/504         4\n",
       "55        625/632        13\n",
       "56        753/760        10\n",
       "57        881/888        16\n",
       "58      1009/1016        12"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attr.df <- attractor2dataframe(attr) \n",
    "attr.df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It would also be useful to have the attractors in 0s and 1s format and know the gene names. Lets make a function for this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "dec2binState <- function(x, genes){ \n",
    "    state <- as.integer( intToBits(x)[1:length(genes)] )  \n",
    "    names(state) <- genes\n",
    "    state\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using this function we can analyze the active nodes in a state. For example, the attractor 58:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li><dl class=dl-horizontal>\n",
       "\t<dt>IL2</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>RORGT</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>STAT3</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>FOXP3</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>TGFB</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL2e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL21e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>TGFBe</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "</dl>\n",
       "</li>\n",
       "\t<li><dl class=dl-horizontal>\n",
       "\t<dt>IL2</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>RORGT</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>STAT3</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>FOXP3</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>TGFB</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL2e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL21e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>TGFBe</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "</dl>\n",
       "</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item \\begin{description*}\n",
       "\\item[IL2] 1\n",
       "\\item[RORGT] 0\n",
       "\\item[STAT3] 0\n",
       "\\item[FOXP3] 0\n",
       "\\item[TGFB] 1\n",
       "\\item[IL10] 1\n",
       "\\item[IL2e] 1\n",
       "\\item[IL21e] 1\n",
       "\\item[TGFBe] 1\n",
       "\\item[IL10e] 1\n",
       "\\end{description*}\n",
       "\n",
       "\\item \\begin{description*}\n",
       "\\item[IL2] 0\n",
       "\\item[RORGT] 0\n",
       "\\item[STAT3] 0\n",
       "\\item[FOXP3] 1\n",
       "\\item[TGFB] 1\n",
       "\\item[IL10] 1\n",
       "\\item[IL2e] 1\n",
       "\\item[IL21e] 1\n",
       "\\item[TGFBe] 1\n",
       "\\item[IL10e] 1\n",
       "\\end{description*}\n",
       "\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. IL2\n",
       ":   1RORGT\n",
       ":   0STAT3\n",
       ":   0FOXP3\n",
       ":   0TGFB\n",
       ":   1IL10\n",
       ":   1IL2e\n",
       ":   1IL21e\n",
       ":   1TGFBe\n",
       ":   1IL10e\n",
       ":   1\n",
       "\n",
       "\n",
       "2. IL2\n",
       ":   0RORGT\n",
       ":   0STAT3\n",
       ":   0FOXP3\n",
       ":   1TGFB\n",
       ":   1IL10\n",
       ":   1IL2e\n",
       ":   1IL21e\n",
       ":   1TGFBe\n",
       ":   1IL10e\n",
       ":   1\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "  IL2 RORGT STAT3 FOXP3  TGFB  IL10  IL2e IL21e TGFBe IL10e \n",
       "    1     0     0     0     1     1     1     1     1     1 \n",
       "\n",
       "[[2]]\n",
       "  IL2 RORGT STAT3 FOXP3  TGFB  IL10  IL2e IL21e TGFBe IL10e \n",
       "    0     0     0     1     1     1     1     1     1     1 \n"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bin.states <- lapply (attr$attractors[[58]]$involvedStates, \n",
    "                      dec2binState, net$genes)\n",
    "bin.states"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Labeling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Attractors correspond to cell types. It is very important to verify that all the expected cell types appear in our attractors, if they are not present we might be missing interactions. It is also important to see if there are attractors that do not correspond to known cell types, as they may be predictions or show errors in the construction of the network.\n",
    "\n",
    "However, when there are a lot of inputs, it is possible that many attractors correspond to a single cell type that can be found in different environments (we will discuss this further). To solve this problem we will label our attractors according to a set of rules."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the case of the Th17/iTreg network this rules are:\n",
    "\n",
    ">labels,  rules\n",
    "\n",
    ">Th0,   ! (RORGT | FOXP3 | TGFB | IL10)\n",
    "\n",
    ">Th17,  RORgt & STAT3\n",
    "\n",
    ">iTreg, FOXP3 & TGFB\n",
    "\n",
    ">IL10+, IL10 & ! (RORGT | FOXP3 | TGFB)\n",
    "\n",
    ">TGFB+, TGFB & ! (RORGT | FOXP3 | TGFB)\n",
    "\n",
    ">IL10+TGFB+, (TGFB & IL10) & ! (RORGT | FOXP3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>labels</th><th scope=col>rules</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Th0</td><td>!(RORGT | FOXP3 | TGFB | IL10)</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Th17</td><td>RORGT & STAT3</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Treg</td><td>FOXP3 & TGFB</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>IL10+</td><td>IL10</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>TGFB+</td><td>TGFB & ! (RORGT | FOXP3)</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       "  & labels & rules\\\\\n",
       "\\hline\n",
       "\t1 & Th0 & !(RORGT | FOXP3 | TGFB | IL10)\\\\\n",
       "\t2 & Th17 & RORGT & STAT3\\\\\n",
       "\t3 & Treg & FOXP3 & TGFB\\\\\n",
       "\t4 & IL10+ & IL10\\\\\n",
       "\t5 & TGFB+ & TGFB & ! (RORGT | FOXP3)\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "  labels                          rules\n",
       "1    Th0 !(RORGT | FOXP3 | TGFB | IL10)\n",
       "2   Th17                  RORGT & STAT3\n",
       "3   Treg                   FOXP3 & TGFB\n",
       "4  IL10+                           IL10\n",
       "5  TGFB+       TGFB & ! (RORGT | FOXP3)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels.rules <- data.frame(\n",
    "    labels = c('Th0', 'Th17', 'Treg', 'IL10+', 'TGFB+'),\n",
    "    rules  = c('!(RORGT | FOXP3 | TGFB | IL10)', \n",
    "            'RORGT & STAT3', \n",
    "            'FOXP3 & TGFB', \n",
    "            'IL10', \n",
    "            'TGFB & ! (RORGT | FOXP3)' ),\n",
    "    stringsAsFactors = FALSE\n",
    ")\n",
    "labels.rules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each rule we can write a function that evaluates  the nodes. For example, for the rule __Th0__ we could write the following rule:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ruleTh0 <- function(\n",
    "    IL2, RORGT, STAT3, FOXP3, TGFB, IL10, IL2e, IL21e, TGFBe, IL10e\n",
    "    ) { \n",
    "    if (!(RORGT | FOXP3 | TGFB | IL10)) 'Th0' \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use this function with a binary state. For examplem the state 0 (0,0,0,0,0,0,0,0,0,0) has the label \"Th0\", but the state 121 (1,0,0,1,1,1,1,,0,0,0) does not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'Th0'"
      ],
      "text/latex": [
       "'Th0'"
      ],
      "text/markdown": [
       "'Th0'"
      ],
      "text/plain": [
       "[1] \"Th0\""
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "NULL"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label <- ruleTh0(0,0,0,0,0,0,0,0,0,0)\n",
    "label\n",
    "state <- ruleTh0(1,0,0,1,1,1,1,0,0,0)\n",
    "state"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also give the function a list of named arguments using __do.call()__, this way the function can receive the state obtained from __dec2binState()__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'Th0'"
      ],
      "text/latex": [
       "'Th0'"
      ],
      "text/markdown": [
       "'Th0'"
      ],
      "text/plain": [
       "[1] \"Th0\""
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "state <- attr$attractors[[1]]$involvedStates\n",
    "state <- dec2binState(state, net$genes)\n",
    "#state\n",
    "label <- do.call(ruleTh0, as.list(state))\n",
    "label"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While we could create specific functions for each labeling rule, it is better to make a general function that can interpret our rules.\n",
    "\n",
    "The function __labelState()__ takes a binary state with gene names, a set of labels and the rules as strings and returns a label."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "labelState <- function(state, node.names, labels, rules) {\n",
    "    names(state) <- node.names\n",
    "    label = c()\n",
    "    for (j in 1:length(rules)) { #evaluate rules\n",
    "        # create string with function\n",
    "        f <- paste('function(', paste(node.names, collapse=','), ') { if (',\n",
    "                   rules[j], ') \\'', labels[j], '\\' }' , sep='')\n",
    "        f <- eval(parse(text=f)) # evaluate string to create function\n",
    "        label <- append(label, do.call(f, as.list(state))) # apply function\n",
    "    }\n",
    "    # format label\n",
    "    if (is.null(label)) { label <-c('NoLabel')}\n",
    "    label <- paste(label, collapse='')\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use this method to check all the possible labels in one go. Some states may have more than one label."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'TregIL10+'"
      ],
      "text/latex": [
       "'TregIL10+'"
      ],
      "text/markdown": [
       "'TregIL10+'"
      ],
      "text/plain": [
       "[1] \"TregIL10+\""
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "state <- dec2binState(121, net$genes)\n",
    "#state\n",
    "label <- labelState(state, net$genes, labels.rules$labels, labels.rules$rules)\n",
    "label"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using this method it is possible to label all the states involved in an attractor. The function __labelAttractors()__ takes an attractors object created by BoolNet and returns a list of the labels for each attractor in order. If an attractor has multiple states it will return a label for each state."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "labelAttractors <- function(attr, node.names, labels, rules) {\n",
    "    attr.labels <- list()\n",
    "    for (i in 1:length(attr$attractors)) {\n",
    "        label <- sapply(attr$attractors[[i]]$involvedStates, function(state) {\n",
    "            state <- dec2binState(state, node.names) #state to binary\n",
    "            l <- labelState(state, node.names, labels, rules) #label\n",
    "        })\n",
    "        attr.labels <- append(attr.labels, list(label))\n",
    "    }\n",
    "    attr.labels\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets obtain the labels of all the attractors using the function __labelAttractors()__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'TGFB+'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'Treg'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'Treg'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "\t<li>'TGFB+'</li>\n",
       "\t<li>'Th17'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'Treg'</li>\n",
       "\t<li>'Th17'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "\t<li>'Th17'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'Treg'</li>\n",
       "\t<li>'Th17'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "\t<li>'IL10+'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'IL10+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'IL10+'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'IL10+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'Th0'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'NoLabel'</li>\n",
       "\t<li>'TGFB+'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'NoLabel'</li>\n",
       "\t<li>'TGFB+'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "</ol>\n",
       "</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item 'Th0'\n",
       "\\item 'Th0'\n",
       "\\item 'Th0'\n",
       "\\item 'TGFB+'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'Th0'\n",
       "\\item 'Th0'\n",
       "\\item 'Treg'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\item 'Th0'\n",
       "\\item 'Th0'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'Th0'\n",
       "\\item 'Th0'\n",
       "\\item 'Treg'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\item 'TGFB+'\n",
       "\\item 'Th17'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'Treg'\n",
       "\\item 'Th17'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\item 'Th17'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'Treg'\n",
       "\\item 'Th17'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\item 'IL10+'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'IL10+'\n",
       "\\item 'TregIL10+'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'IL10+'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'IL10+'\n",
       "\\item 'TregIL10+'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'Th0'\n",
       "\\item 'Th0'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'NoLabel'\n",
       "\\item 'TGFB+'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'NoLabel'\n",
       "\\item 'TGFB+'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'TregIL10+'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. 'Th0'\n",
       "2. 'Th0'\n",
       "3. 'Th0'\n",
       "4. 'TGFB+'\n",
       "5. 'IL10+TGFB+'\n",
       "6. 'Th0'\n",
       "7. 'Th0'\n",
       "8. 'Treg'\n",
       "9. 'IL10+TGFB+'\n",
       "10. 'TregIL10+'\n",
       "11. 'Th0'\n",
       "12. 'Th0'\n",
       "13. 'IL10+TGFB+'\n",
       "14. 'Th0'\n",
       "15. 'Th0'\n",
       "16. 'Treg'\n",
       "17. 'IL10+TGFB+'\n",
       "18. 'TregIL10+'\n",
       "19. 'TGFB+'\n",
       "20. 'Th17'\n",
       "21. 'IL10+TGFB+'\n",
       "22. 'Treg'\n",
       "23. 'Th17'\n",
       "24. 'IL10+TGFB+'\n",
       "25. 'TregIL10+'\n",
       "26. 'Th17'\n",
       "27. 'IL10+TGFB+'\n",
       "28. 'Treg'\n",
       "29. 'Th17'\n",
       "30. 'IL10+TGFB+'\n",
       "31. 'TregIL10+'\n",
       "32. 'IL10+'\n",
       "33. 'IL10+TGFB+'\n",
       "34. 'IL10+'\n",
       "35. 'TregIL10+'\n",
       "36. 'IL10+TGFB+'\n",
       "37. 'IL10+'\n",
       "38. 'IL10+TGFB+'\n",
       "39. 'IL10+'\n",
       "40. 'TregIL10+'\n",
       "41. 'IL10+TGFB+'\n",
       "42. 'IL10+TGFB+'\n",
       "43. 'IL10+TGFB+'\n",
       "44. 'TregIL10+'\n",
       "45. 'IL10+TGFB+'\n",
       "46. 'IL10+TGFB+'\n",
       "47. 'TregIL10+'\n",
       "48. 1. 'IL10+TGFB+'\n",
       "2. 'TregIL10+'\n",
       "\n",
       "\n",
       "\n",
       "49. 1. 'Th0'\n",
       "2. 'Th0'\n",
       "\n",
       "\n",
       "\n",
       "50. 1. 'IL10+TGFB+'\n",
       "2. 'TregIL10+'\n",
       "\n",
       "\n",
       "\n",
       "51. 1. 'NoLabel'\n",
       "2. 'TGFB+'\n",
       "\n",
       "\n",
       "\n",
       "52. 1. 'IL10+TGFB+'\n",
       "2. 'TregIL10+'\n",
       "\n",
       "\n",
       "\n",
       "53. 1. 'NoLabel'\n",
       "2. 'TGFB+'\n",
       "\n",
       "\n",
       "\n",
       "54. 1. 'IL10+TGFB+'\n",
       "2. 'TregIL10+'\n",
       "\n",
       "\n",
       "\n",
       "55. 1. 'IL10+TGFB+'\n",
       "2. 'TregIL10+'\n",
       "\n",
       "\n",
       "\n",
       "56. 1. 'IL10+TGFB+'\n",
       "2. 'TregIL10+'\n",
       "\n",
       "\n",
       "\n",
       "57. 1. 'IL10+TGFB+'\n",
       "2. 'TregIL10+'\n",
       "\n",
       "\n",
       "\n",
       "58. 1. 'IL10+TGFB+'\n",
       "2. 'TregIL10+'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] \"Th0\"\n",
       "\n",
       "[[2]]\n",
       "[1] \"Th0\"\n",
       "\n",
       "[[3]]\n",
       "[1] \"Th0\"\n",
       "\n",
       "[[4]]\n",
       "[1] \"TGFB+\"\n",
       "\n",
       "[[5]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[6]]\n",
       "[1] \"Th0\"\n",
       "\n",
       "[[7]]\n",
       "[1] \"Th0\"\n",
       "\n",
       "[[8]]\n",
       "[1] \"Treg\"\n",
       "\n",
       "[[9]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[10]]\n",
       "[1] \"TregIL10+\"\n",
       "\n",
       "[[11]]\n",
       "[1] \"Th0\"\n",
       "\n",
       "[[12]]\n",
       "[1] \"Th0\"\n",
       "\n",
       "[[13]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[14]]\n",
       "[1] \"Th0\"\n",
       "\n",
       "[[15]]\n",
       "[1] \"Th0\"\n",
       "\n",
       "[[16]]\n",
       "[1] \"Treg\"\n",
       "\n",
       "[[17]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[18]]\n",
       "[1] \"TregIL10+\"\n",
       "\n",
       "[[19]]\n",
       "[1] \"TGFB+\"\n",
       "\n",
       "[[20]]\n",
       "[1] \"Th17\"\n",
       "\n",
       "[[21]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[22]]\n",
       "[1] \"Treg\"\n",
       "\n",
       "[[23]]\n",
       "[1] \"Th17\"\n",
       "\n",
       "[[24]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[25]]\n",
       "[1] \"TregIL10+\"\n",
       "\n",
       "[[26]]\n",
       "[1] \"Th17\"\n",
       "\n",
       "[[27]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[28]]\n",
       "[1] \"Treg\"\n",
       "\n",
       "[[29]]\n",
       "[1] \"Th17\"\n",
       "\n",
       "[[30]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[31]]\n",
       "[1] \"TregIL10+\"\n",
       "\n",
       "[[32]]\n",
       "[1] \"IL10+\"\n",
       "\n",
       "[[33]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[34]]\n",
       "[1] \"IL10+\"\n",
       "\n",
       "[[35]]\n",
       "[1] \"TregIL10+\"\n",
       "\n",
       "[[36]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[37]]\n",
       "[1] \"IL10+\"\n",
       "\n",
       "[[38]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[39]]\n",
       "[1] \"IL10+\"\n",
       "\n",
       "[[40]]\n",
       "[1] \"TregIL10+\"\n",
       "\n",
       "[[41]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[42]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[43]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[44]]\n",
       "[1] \"TregIL10+\"\n",
       "\n",
       "[[45]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[46]]\n",
       "[1] \"IL10+TGFB+\"\n",
       "\n",
       "[[47]]\n",
       "[1] \"TregIL10+\"\n",
       "\n",
       "[[48]]\n",
       "[1] \"IL10+TGFB+\" \"TregIL10+\" \n",
       "\n",
       "[[49]]\n",
       "[1] \"Th0\" \"Th0\"\n",
       "\n",
       "[[50]]\n",
       "[1] \"IL10+TGFB+\" \"TregIL10+\" \n",
       "\n",
       "[[51]]\n",
       "[1] \"NoLabel\" \"TGFB+\"  \n",
       "\n",
       "[[52]]\n",
       "[1] \"IL10+TGFB+\" \"TregIL10+\" \n",
       "\n",
       "[[53]]\n",
       "[1] \"NoLabel\" \"TGFB+\"  \n",
       "\n",
       "[[54]]\n",
       "[1] \"IL10+TGFB+\" \"TregIL10+\" \n",
       "\n",
       "[[55]]\n",
       "[1] \"IL10+TGFB+\" \"TregIL10+\" \n",
       "\n",
       "[[56]]\n",
       "[1] \"IL10+TGFB+\" \"TregIL10+\" \n",
       "\n",
       "[[57]]\n",
       "[1] \"IL10+TGFB+\" \"TregIL10+\" \n",
       "\n",
       "[[58]]\n",
       "[1] \"IL10+TGFB+\" \"TregIL10+\" \n"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels <- labelAttractors(attr, net$genes, labels.rules$labels, labels.rules$rules)\n",
    "labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will add the labels to the dataframe we created with __attractor2dataframe()__. To do this we will need to join the labels of the attractors that have more than one state with \"/\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>involvedStates</th><th scope=col>basinSize</th><th scope=col>label</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>0</td><td>27</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td>2</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>4</td><td>14</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>16</td><td>13</td><td>TGFB+</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>48</td><td>8</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>65</td><td>35</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>68</td><td>8</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>89</td><td>13</td><td>Treg</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>112</td><td>3</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>121</td><td>1</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>132</td><td>54</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>129</td><td>2</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>176</td><td>8</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>193</td><td>2</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>196</td><td>8</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>217</td><td>10</td><td>Treg</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>240</td><td>3</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>249</td><td>1</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>272</td><td>26</td><td>TGFB+</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>278</td><td>14</td><td>Th17</td></tr>\n",
       "\t<tr><th scope=row>21</th><td>304</td><td>24</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>22</th><td>345</td><td>26</td><td>Treg</td></tr>\n",
       "\t<tr><th scope=row>23</th><td>342</td><td>8</td><td>Th17</td></tr>\n",
       "\t<tr><th scope=row>24</th><td>368</td><td>19</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>25</th><td>377</td><td>1</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>26</th><td>406</td><td>40</td><td>Th17</td></tr>\n",
       "\t<tr><th scope=row>27</th><td>432</td><td>24</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>28</th><td>473</td><td>12</td><td>Treg</td></tr>\n",
       "\t<tr><th scope=row>29</th><td>470</td><td>8</td><td>Th17</td></tr>\n",
       "\t<tr><th scope=row>30</th><td>496</td><td>19</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>31</th><td>505</td><td>1</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>32</th><td>544</td><td>43</td><td>IL10+</td></tr>\n",
       "\t<tr><th scope=row>33</th><td>560</td><td>21</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>34</th><td>608</td><td>43</td><td>IL10+</td></tr>\n",
       "\t<tr><th scope=row>35</th><td>633</td><td>5</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>36</th><td>624</td><td>3</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>37</th><td>672</td><td>46</td><td>IL10+</td></tr>\n",
       "\t<tr><th scope=row>38</th><td>688</td><td>18</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>39</th><td>736</td><td>46</td><td>IL10+</td></tr>\n",
       "\t<tr><th scope=row>40</th><td>761</td><td>5</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>41</th><td>752</td><td>3</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>42</th><td>816</td><td>64</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>43</th><td>880</td><td>43</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>44</th><td>889</td><td>5</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>45</th><td>944</td><td>64</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>46</th><td>1008</td><td>47</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>47</th><td>1017</td><td>5</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>48</th><td>113/120</td><td>4</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>49</th><td>192/197</td><td>36</td><td>Th0/Th0</td></tr>\n",
       "\t<tr><th scope=row>50</th><td>241/248</td><td>4</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>51</th><td>338/341</td><td>6</td><td>NoLabel/TGFB+</td></tr>\n",
       "\t<tr><th scope=row>52</th><td>369/376</td><td>4</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>53</th><td>466/469</td><td>20</td><td>NoLabel/TGFB+</td></tr>\n",
       "\t<tr><th scope=row>54</th><td>497/504</td><td>4</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>55</th><td>625/632</td><td>13</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>56</th><td>753/760</td><td>10</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>57</th><td>881/888</td><td>16</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>58</th><td>1009/1016</td><td>12</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lll}\n",
       "  & involvedStates & basinSize & label\\\\\n",
       "\\hline\n",
       "\t1 & 0 & 27 & Th0\\\\\n",
       "\t2 & 1 & 2 & Th0\\\\\n",
       "\t3 & 4 & 14 & Th0\\\\\n",
       "\t4 & 16 & 13 & TGFB+\\\\\n",
       "\t5 & 48 & 8 & IL10+TGFB+\\\\\n",
       "\t6 & 65 & 35 & Th0\\\\\n",
       "\t7 & 68 & 8 & Th0\\\\\n",
       "\t8 & 89 & 13 & Treg\\\\\n",
       "\t9 & 112 & 3 & IL10+TGFB+\\\\\n",
       "\t10 & 121 & 1 & TregIL10+\\\\\n",
       "\t11 & 132 & 54 & Th0\\\\\n",
       "\t12 & 129 & 2 & Th0\\\\\n",
       "\t13 & 176 & 8 & IL10+TGFB+\\\\\n",
       "\t14 & 193 & 2 & Th0\\\\\n",
       "\t15 & 196 & 8 & Th0\\\\\n",
       "\t16 & 217 & 10 & Treg\\\\\n",
       "\t17 & 240 & 3 & IL10+TGFB+\\\\\n",
       "\t18 & 249 & 1 & TregIL10+\\\\\n",
       "\t19 & 272 & 26 & TGFB+\\\\\n",
       "\t20 & 278 & 14 & Th17\\\\\n",
       "\t21 & 304 & 24 & IL10+TGFB+\\\\\n",
       "\t22 & 345 & 26 & Treg\\\\\n",
       "\t23 & 342 & 8 & Th17\\\\\n",
       "\t24 & 368 & 19 & IL10+TGFB+\\\\\n",
       "\t25 & 377 & 1 & TregIL10+\\\\\n",
       "\t26 & 406 & 40 & Th17\\\\\n",
       "\t27 & 432 & 24 & IL10+TGFB+\\\\\n",
       "\t28 & 473 & 12 & Treg\\\\\n",
       "\t29 & 470 & 8 & Th17\\\\\n",
       "\t30 & 496 & 19 & IL10+TGFB+\\\\\n",
       "\t31 & 505 & 1 & TregIL10+\\\\\n",
       "\t32 & 544 & 43 & IL10+\\\\\n",
       "\t33 & 560 & 21 & IL10+TGFB+\\\\\n",
       "\t34 & 608 & 43 & IL10+\\\\\n",
       "\t35 & 633 & 5 & TregIL10+\\\\\n",
       "\t36 & 624 & 3 & IL10+TGFB+\\\\\n",
       "\t37 & 672 & 46 & IL10+\\\\\n",
       "\t38 & 688 & 18 & IL10+TGFB+\\\\\n",
       "\t39 & 736 & 46 & IL10+\\\\\n",
       "\t40 & 761 & 5 & TregIL10+\\\\\n",
       "\t41 & 752 & 3 & IL10+TGFB+\\\\\n",
       "\t42 & 816 & 64 & IL10+TGFB+\\\\\n",
       "\t43 & 880 & 43 & IL10+TGFB+\\\\\n",
       "\t44 & 889 & 5 & TregIL10+\\\\\n",
       "\t45 & 944 & 64 & IL10+TGFB+\\\\\n",
       "\t46 & 1008 & 47 & IL10+TGFB+\\\\\n",
       "\t47 & 1017 & 5 & TregIL10+\\\\\n",
       "\t48 & 113/120 & 4 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t49 & 192/197 & 36 & Th0/Th0\\\\\n",
       "\t50 & 241/248 & 4 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t51 & 338/341 & 6 & NoLabel/TGFB+\\\\\n",
       "\t52 & 369/376 & 4 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t53 & 466/469 & 20 & NoLabel/TGFB+\\\\\n",
       "\t54 & 497/504 & 4 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t55 & 625/632 & 13 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t56 & 753/760 & 10 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t57 & 881/888 & 16 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t58 & 1009/1016 & 12 & IL10+TGFB+/TregIL10+\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "   involvedStates basinSize                label\n",
       "1               0        27                  Th0\n",
       "2               1         2                  Th0\n",
       "3               4        14                  Th0\n",
       "4              16        13                TGFB+\n",
       "5              48         8           IL10+TGFB+\n",
       "6              65        35                  Th0\n",
       "7              68         8                  Th0\n",
       "8              89        13                 Treg\n",
       "9             112         3           IL10+TGFB+\n",
       "10            121         1            TregIL10+\n",
       "11            132        54                  Th0\n",
       "12            129         2                  Th0\n",
       "13            176         8           IL10+TGFB+\n",
       "14            193         2                  Th0\n",
       "15            196         8                  Th0\n",
       "16            217        10                 Treg\n",
       "17            240         3           IL10+TGFB+\n",
       "18            249         1            TregIL10+\n",
       "19            272        26                TGFB+\n",
       "20            278        14                 Th17\n",
       "21            304        24           IL10+TGFB+\n",
       "22            345        26                 Treg\n",
       "23            342         8                 Th17\n",
       "24            368        19           IL10+TGFB+\n",
       "25            377         1            TregIL10+\n",
       "26            406        40                 Th17\n",
       "27            432        24           IL10+TGFB+\n",
       "28            473        12                 Treg\n",
       "29            470         8                 Th17\n",
       "30            496        19           IL10+TGFB+\n",
       "31            505         1            TregIL10+\n",
       "32            544        43                IL10+\n",
       "33            560        21           IL10+TGFB+\n",
       "34            608        43                IL10+\n",
       "35            633         5            TregIL10+\n",
       "36            624         3           IL10+TGFB+\n",
       "37            672        46                IL10+\n",
       "38            688        18           IL10+TGFB+\n",
       "39            736        46                IL10+\n",
       "40            761         5            TregIL10+\n",
       "41            752         3           IL10+TGFB+\n",
       "42            816        64           IL10+TGFB+\n",
       "43            880        43           IL10+TGFB+\n",
       "44            889         5            TregIL10+\n",
       "45            944        64           IL10+TGFB+\n",
       "46           1008        47           IL10+TGFB+\n",
       "47           1017         5            TregIL10+\n",
       "48        113/120         4 IL10+TGFB+/TregIL10+\n",
       "49        192/197        36              Th0/Th0\n",
       "50        241/248         4 IL10+TGFB+/TregIL10+\n",
       "51        338/341         6        NoLabel/TGFB+\n",
       "52        369/376         4 IL10+TGFB+/TregIL10+\n",
       "53        466/469        20        NoLabel/TGFB+\n",
       "54        497/504         4 IL10+TGFB+/TregIL10+\n",
       "55        625/632        13 IL10+TGFB+/TregIL10+\n",
       "56        753/760        10 IL10+TGFB+/TregIL10+\n",
       "57        881/888        16 IL10+TGFB+/TregIL10+\n",
       "58      1009/1016        12 IL10+TGFB+/TregIL10+"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attr.df$label <- sapply(labels, function(label) {\n",
    "    paste(as.character(label), collapse='/')\n",
    "})\n",
    "attr.df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see some labels have more than one attractor. This is because some cell types can be found in different environments. We can see all the unique labels to determine if all the labels correspond to the expected cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'Th0'</li>\n",
       "\t<li>'TGFB+'</li>\n",
       "\t<li>'IL10+TGFB+'</li>\n",
       "\t<li>'Treg'</li>\n",
       "\t<li>'TregIL10+'</li>\n",
       "\t<li>'Th17'</li>\n",
       "\t<li>'IL10+'</li>\n",
       "\t<li>'IL10+TGFB+/TregIL10+'</li>\n",
       "\t<li>'Th0/Th0'</li>\n",
       "\t<li>'NoLabel/TGFB+'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Th0'\n",
       "\\item 'TGFB+'\n",
       "\\item 'IL10+TGFB+'\n",
       "\\item 'Treg'\n",
       "\\item 'TregIL10+'\n",
       "\\item 'Th17'\n",
       "\\item 'IL10+'\n",
       "\\item 'IL10+TGFB+/TregIL10+'\n",
       "\\item 'Th0/Th0'\n",
       "\\item 'NoLabel/TGFB+'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Th0'\n",
       "2. 'TGFB+'\n",
       "3. 'IL10+TGFB+'\n",
       "4. 'Treg'\n",
       "5. 'TregIL10+'\n",
       "6. 'Th17'\n",
       "7. 'IL10+'\n",
       "8. 'IL10+TGFB+/TregIL10+'\n",
       "9. 'Th0/Th0'\n",
       "10. 'NoLabel/TGFB+'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"Th0\"                  \"TGFB+\"                \"IL10+TGFB+\"          \n",
       " [4] \"Treg\"                 \"TregIL10+\"            \"Th17\"                \n",
       " [7] \"IL10+\"                \"IL10+TGFB+/TregIL10+\" \"Th0/Th0\"             \n",
       "[10] \"NoLabel/TGFB+\"       "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique(attr.df$label)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While we have all the expected cell types we have some cycles where one of the states could not be labeled. If we check the dataframe, the states 338 and 366 have no label, we can study them to see what happened."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl class=dl-horizontal>\n",
       "\t<dt>IL2</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>RORGT</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>STAT3</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>FOXP3</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>TGFB</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>IL2e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL21e</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>TGFBe</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10e</dt>\n",
       "\t\t<dd>0</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[IL2] 0\n",
       "\\item[RORGT] 1\n",
       "\\item[STAT3] 0\n",
       "\\item[FOXP3] 0\n",
       "\\item[TGFB] 1\n",
       "\\item[IL10] 0\n",
       "\\item[IL2e] 1\n",
       "\\item[IL21e] 0\n",
       "\\item[TGFBe] 1\n",
       "\\item[IL10e] 0\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "IL2\n",
       ":   0RORGT\n",
       ":   1STAT3\n",
       ":   0FOXP3\n",
       ":   0TGFB\n",
       ":   1IL10\n",
       ":   0IL2e\n",
       ":   1IL21e\n",
       ":   0TGFBe\n",
       ":   1IL10e\n",
       ":   0\n",
       "\n"
      ],
      "text/plain": [
       "  IL2 RORGT STAT3 FOXP3  TGFB  IL10  IL2e IL21e TGFBe IL10e \n",
       "    0     1     0     0     1     0     1     0     1     0 "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<dl class=dl-horizontal>\n",
       "\t<dt>IL2</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>RORGT</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>STAT3</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>FOXP3</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>TGFB</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>IL2e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL21e</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>TGFBe</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10e</dt>\n",
       "\t\t<dd>0</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[IL2] 1\n",
       "\\item[RORGT] 0\n",
       "\\item[STAT3] 1\n",
       "\\item[FOXP3] 0\n",
       "\\item[TGFB] 1\n",
       "\\item[IL10] 0\n",
       "\\item[IL2e] 1\n",
       "\\item[IL21e] 0\n",
       "\\item[TGFBe] 1\n",
       "\\item[IL10e] 0\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "IL2\n",
       ":   1RORGT\n",
       ":   0STAT3\n",
       ":   1FOXP3\n",
       ":   0TGFB\n",
       ":   1IL10\n",
       ":   0IL2e\n",
       ":   1IL21e\n",
       ":   0TGFBe\n",
       ":   1IL10e\n",
       ":   0\n",
       "\n"
      ],
      "text/plain": [
       "  IL2 RORGT STAT3 FOXP3  TGFB  IL10  IL2e IL21e TGFBe IL10e \n",
       "    1     0     1     0     1     0     1     0     1     0 "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<dl class=dl-horizontal>\n",
       "\t<dt>IL2</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>RORGT</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>STAT3</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>FOXP3</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>TGFB</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>IL2e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL21e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>TGFBe</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10e</dt>\n",
       "\t\t<dd>0</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[IL2] 0\n",
       "\\item[RORGT] 1\n",
       "\\item[STAT3] 0\n",
       "\\item[FOXP3] 0\n",
       "\\item[TGFB] 1\n",
       "\\item[IL10] 0\n",
       "\\item[IL2e] 1\n",
       "\\item[IL21e] 1\n",
       "\\item[TGFBe] 1\n",
       "\\item[IL10e] 0\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "IL2\n",
       ":   0RORGT\n",
       ":   1STAT3\n",
       ":   0FOXP3\n",
       ":   0TGFB\n",
       ":   1IL10\n",
       ":   0IL2e\n",
       ":   1IL21e\n",
       ":   1TGFBe\n",
       ":   1IL10e\n",
       ":   0\n",
       "\n"
      ],
      "text/plain": [
       "  IL2 RORGT STAT3 FOXP3  TGFB  IL10  IL2e IL21e TGFBe IL10e \n",
       "    0     1     0     0     1     0     1     1     1     0 "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<dl class=dl-horizontal>\n",
       "\t<dt>IL2</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>RORGT</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>STAT3</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>FOXP3</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>TGFB</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10</dt>\n",
       "\t\t<dd>0</dd>\n",
       "\t<dt>IL2e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL21e</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>TGFBe</dt>\n",
       "\t\t<dd>1</dd>\n",
       "\t<dt>IL10e</dt>\n",
       "\t\t<dd>0</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[IL2] 1\n",
       "\\item[RORGT] 0\n",
       "\\item[STAT3] 1\n",
       "\\item[FOXP3] 0\n",
       "\\item[TGFB] 1\n",
       "\\item[IL10] 0\n",
       "\\item[IL2e] 1\n",
       "\\item[IL21e] 1\n",
       "\\item[TGFBe] 1\n",
       "\\item[IL10e] 0\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "IL2\n",
       ":   1RORGT\n",
       ":   0STAT3\n",
       ":   1FOXP3\n",
       ":   0TGFB\n",
       ":   1IL10\n",
       ":   0IL2e\n",
       ":   1IL21e\n",
       ":   1TGFBe\n",
       ":   1IL10e\n",
       ":   0\n",
       "\n"
      ],
      "text/plain": [
       "  IL2 RORGT STAT3 FOXP3  TGFB  IL10  IL2e IL21e TGFBe IL10e \n",
       "    1     0     1     0     1     0     1     1     1     0 "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "state.odd.a1 <- dec2binState(338, net$genes)\n",
    "state.odd.a2 <- dec2binState(341, net$genes)\n",
    "state.odd.a1\n",
    "state.odd.a2\n",
    "state.odd.b1 <- dec2binState(466, net$genes)\n",
    "state.odd.b2 <- dec2binState(469, net$genes)\n",
    "state.odd.b1\n",
    "state.odd.b2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see this states correspond to oscillations where ROR$\\gamma$t can not stabilish a stable expression. If we use an asynchronous update this states will dissapear. \n",
    "\n",
    "We can create a new rule to include this states.\n",
    "> RORGT+, RORGT & ! STAT3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>labels</th><th scope=col>rules</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Th0</td><td>!(RORGT | FOXP3 | TGFB | IL10)</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Th17</td><td>RORGT & STAT3</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Treg</td><td>FOXP3 & TGFB</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>IL10+</td><td>IL10</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>TGFB+</td><td>TGFB & ! (RORGT | FOXP3)</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>RORGT+</td><td>RORGT & ! STAT3</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       "  & labels & rules\\\\\n",
       "\\hline\n",
       "\t1 & Th0 & !(RORGT | FOXP3 | TGFB | IL10)\\\\\n",
       "\t2 & Th17 & RORGT & STAT3\\\\\n",
       "\t3 & Treg & FOXP3 & TGFB\\\\\n",
       "\t4 & IL10+ & IL10\\\\\n",
       "\t5 & TGFB+ & TGFB & ! (RORGT | FOXP3)\\\\\n",
       "\t6 & RORGT+ & RORGT & ! STAT3\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "  labels                          rules\n",
       "1    Th0 !(RORGT | FOXP3 | TGFB | IL10)\n",
       "2   Th17                  RORGT & STAT3\n",
       "3   Treg                   FOXP3 & TGFB\n",
       "4  IL10+                           IL10\n",
       "5  TGFB+       TGFB & ! (RORGT | FOXP3)\n",
       "6 RORGT+                RORGT & ! STAT3"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels.rules <- rbind(labels.rules, c('RORGT+', 'RORGT & ! STAT3'))\n",
    "labels.rules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will need to do the labeling and the dataframe once more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>involvedStates</th><th scope=col>basinSize</th><th scope=col>label</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>0</td><td>27</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td>2</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>4</td><td>14</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>16</td><td>13</td><td>TGFB+</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>48</td><td>8</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>65</td><td>35</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>68</td><td>8</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>89</td><td>13</td><td>Treg</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>112</td><td>3</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>121</td><td>1</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>132</td><td>54</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>129</td><td>2</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>176</td><td>8</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>193</td><td>2</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>196</td><td>8</td><td>Th0</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>217</td><td>10</td><td>Treg</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>240</td><td>3</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>249</td><td>1</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>272</td><td>26</td><td>TGFB+</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>278</td><td>14</td><td>Th17</td></tr>\n",
       "\t<tr><th scope=row>21</th><td>304</td><td>24</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>22</th><td>345</td><td>26</td><td>Treg</td></tr>\n",
       "\t<tr><th scope=row>23</th><td>342</td><td>8</td><td>Th17</td></tr>\n",
       "\t<tr><th scope=row>24</th><td>368</td><td>19</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>25</th><td>377</td><td>1</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>26</th><td>406</td><td>40</td><td>Th17</td></tr>\n",
       "\t<tr><th scope=row>27</th><td>432</td><td>24</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>28</th><td>473</td><td>12</td><td>Treg</td></tr>\n",
       "\t<tr><th scope=row>29</th><td>470</td><td>8</td><td>Th17</td></tr>\n",
       "\t<tr><th scope=row>30</th><td>496</td><td>19</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>31</th><td>505</td><td>1</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>32</th><td>544</td><td>43</td><td>IL10+</td></tr>\n",
       "\t<tr><th scope=row>33</th><td>560</td><td>21</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>34</th><td>608</td><td>43</td><td>IL10+</td></tr>\n",
       "\t<tr><th scope=row>35</th><td>633</td><td>5</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>36</th><td>624</td><td>3</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>37</th><td>672</td><td>46</td><td>IL10+</td></tr>\n",
       "\t<tr><th scope=row>38</th><td>688</td><td>18</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>39</th><td>736</td><td>46</td><td>IL10+</td></tr>\n",
       "\t<tr><th scope=row>40</th><td>761</td><td>5</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>41</th><td>752</td><td>3</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>42</th><td>816</td><td>64</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>43</th><td>880</td><td>43</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>44</th><td>889</td><td>5</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>45</th><td>944</td><td>64</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>46</th><td>1008</td><td>47</td><td>IL10+TGFB+</td></tr>\n",
       "\t<tr><th scope=row>47</th><td>1017</td><td>5</td><td>TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>48</th><td>113/120</td><td>4</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>49</th><td>192/197</td><td>36</td><td>Th0/Th0</td></tr>\n",
       "\t<tr><th scope=row>50</th><td>241/248</td><td>4</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>51</th><td>338/341</td><td>6</td><td>RORGT+/TGFB+</td></tr>\n",
       "\t<tr><th scope=row>52</th><td>369/376</td><td>4</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>53</th><td>466/469</td><td>20</td><td>RORGT+/TGFB+</td></tr>\n",
       "\t<tr><th scope=row>54</th><td>497/504</td><td>4</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>55</th><td>625/632</td><td>13</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>56</th><td>753/760</td><td>10</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>57</th><td>881/888</td><td>16</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "\t<tr><th scope=row>58</th><td>1009/1016</td><td>12</td><td>IL10+TGFB+/TregIL10+</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lll}\n",
       "  & involvedStates & basinSize & label\\\\\n",
       "\\hline\n",
       "\t1 & 0 & 27 & Th0\\\\\n",
       "\t2 & 1 & 2 & Th0\\\\\n",
       "\t3 & 4 & 14 & Th0\\\\\n",
       "\t4 & 16 & 13 & TGFB+\\\\\n",
       "\t5 & 48 & 8 & IL10+TGFB+\\\\\n",
       "\t6 & 65 & 35 & Th0\\\\\n",
       "\t7 & 68 & 8 & Th0\\\\\n",
       "\t8 & 89 & 13 & Treg\\\\\n",
       "\t9 & 112 & 3 & IL10+TGFB+\\\\\n",
       "\t10 & 121 & 1 & TregIL10+\\\\\n",
       "\t11 & 132 & 54 & Th0\\\\\n",
       "\t12 & 129 & 2 & Th0\\\\\n",
       "\t13 & 176 & 8 & IL10+TGFB+\\\\\n",
       "\t14 & 193 & 2 & Th0\\\\\n",
       "\t15 & 196 & 8 & Th0\\\\\n",
       "\t16 & 217 & 10 & Treg\\\\\n",
       "\t17 & 240 & 3 & IL10+TGFB+\\\\\n",
       "\t18 & 249 & 1 & TregIL10+\\\\\n",
       "\t19 & 272 & 26 & TGFB+\\\\\n",
       "\t20 & 278 & 14 & Th17\\\\\n",
       "\t21 & 304 & 24 & IL10+TGFB+\\\\\n",
       "\t22 & 345 & 26 & Treg\\\\\n",
       "\t23 & 342 & 8 & Th17\\\\\n",
       "\t24 & 368 & 19 & IL10+TGFB+\\\\\n",
       "\t25 & 377 & 1 & TregIL10+\\\\\n",
       "\t26 & 406 & 40 & Th17\\\\\n",
       "\t27 & 432 & 24 & IL10+TGFB+\\\\\n",
       "\t28 & 473 & 12 & Treg\\\\\n",
       "\t29 & 470 & 8 & Th17\\\\\n",
       "\t30 & 496 & 19 & IL10+TGFB+\\\\\n",
       "\t31 & 505 & 1 & TregIL10+\\\\\n",
       "\t32 & 544 & 43 & IL10+\\\\\n",
       "\t33 & 560 & 21 & IL10+TGFB+\\\\\n",
       "\t34 & 608 & 43 & IL10+\\\\\n",
       "\t35 & 633 & 5 & TregIL10+\\\\\n",
       "\t36 & 624 & 3 & IL10+TGFB+\\\\\n",
       "\t37 & 672 & 46 & IL10+\\\\\n",
       "\t38 & 688 & 18 & IL10+TGFB+\\\\\n",
       "\t39 & 736 & 46 & IL10+\\\\\n",
       "\t40 & 761 & 5 & TregIL10+\\\\\n",
       "\t41 & 752 & 3 & IL10+TGFB+\\\\\n",
       "\t42 & 816 & 64 & IL10+TGFB+\\\\\n",
       "\t43 & 880 & 43 & IL10+TGFB+\\\\\n",
       "\t44 & 889 & 5 & TregIL10+\\\\\n",
       "\t45 & 944 & 64 & IL10+TGFB+\\\\\n",
       "\t46 & 1008 & 47 & IL10+TGFB+\\\\\n",
       "\t47 & 1017 & 5 & TregIL10+\\\\\n",
       "\t48 & 113/120 & 4 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t49 & 192/197 & 36 & Th0/Th0\\\\\n",
       "\t50 & 241/248 & 4 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t51 & 338/341 & 6 & RORGT+/TGFB+\\\\\n",
       "\t52 & 369/376 & 4 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t53 & 466/469 & 20 & RORGT+/TGFB+\\\\\n",
       "\t54 & 497/504 & 4 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t55 & 625/632 & 13 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t56 & 753/760 & 10 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t57 & 881/888 & 16 & IL10+TGFB+/TregIL10+\\\\\n",
       "\t58 & 1009/1016 & 12 & IL10+TGFB+/TregIL10+\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "   involvedStates basinSize                label\n",
       "1               0        27                  Th0\n",
       "2               1         2                  Th0\n",
       "3               4        14                  Th0\n",
       "4              16        13                TGFB+\n",
       "5              48         8           IL10+TGFB+\n",
       "6              65        35                  Th0\n",
       "7              68         8                  Th0\n",
       "8              89        13                 Treg\n",
       "9             112         3           IL10+TGFB+\n",
       "10            121         1            TregIL10+\n",
       "11            132        54                  Th0\n",
       "12            129         2                  Th0\n",
       "13            176         8           IL10+TGFB+\n",
       "14            193         2                  Th0\n",
       "15            196         8                  Th0\n",
       "16            217        10                 Treg\n",
       "17            240         3           IL10+TGFB+\n",
       "18            249         1            TregIL10+\n",
       "19            272        26                TGFB+\n",
       "20            278        14                 Th17\n",
       "21            304        24           IL10+TGFB+\n",
       "22            345        26                 Treg\n",
       "23            342         8                 Th17\n",
       "24            368        19           IL10+TGFB+\n",
       "25            377         1            TregIL10+\n",
       "26            406        40                 Th17\n",
       "27            432        24           IL10+TGFB+\n",
       "28            473        12                 Treg\n",
       "29            470         8                 Th17\n",
       "30            496        19           IL10+TGFB+\n",
       "31            505         1            TregIL10+\n",
       "32            544        43                IL10+\n",
       "33            560        21           IL10+TGFB+\n",
       "34            608        43                IL10+\n",
       "35            633         5            TregIL10+\n",
       "36            624         3           IL10+TGFB+\n",
       "37            672        46                IL10+\n",
       "38            688        18           IL10+TGFB+\n",
       "39            736        46                IL10+\n",
       "40            761         5            TregIL10+\n",
       "41            752         3           IL10+TGFB+\n",
       "42            816        64           IL10+TGFB+\n",
       "43            880        43           IL10+TGFB+\n",
       "44            889         5            TregIL10+\n",
       "45            944        64           IL10+TGFB+\n",
       "46           1008        47           IL10+TGFB+\n",
       "47           1017         5            TregIL10+\n",
       "48        113/120         4 IL10+TGFB+/TregIL10+\n",
       "49        192/197        36              Th0/Th0\n",
       "50        241/248         4 IL10+TGFB+/TregIL10+\n",
       "51        338/341         6         RORGT+/TGFB+\n",
       "52        369/376         4 IL10+TGFB+/TregIL10+\n",
       "53        466/469        20         RORGT+/TGFB+\n",
       "54        497/504         4 IL10+TGFB+/TregIL10+\n",
       "55        625/632        13 IL10+TGFB+/TregIL10+\n",
       "56        753/760        10 IL10+TGFB+/TregIL10+\n",
       "57        881/888        16 IL10+TGFB+/TregIL10+\n",
       "58      1009/1016        12 IL10+TGFB+/TregIL10+"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels <- list()\n",
    "for (i in 1:length(attr$attractors)) {\n",
    "    label <- sapply(attr$attractors[[i]]$involvedStates, function(state) {\n",
    "        state <- dec2binState(state, net$genes)\n",
    "        labelState(state,  net$genes, labels.rules$labels, labels.rules$rules)\n",
    "    })\n",
    "    labels <- append(labels, list(label))\n",
    "}\n",
    "\n",
    "attr.df$label <- sapply(labels, function(label) {\n",
    "    paste(as.character(label), collapse='/')\n",
    "})\n",
    "attr.df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that all the states are labeled we can also determine which attractor states correspond to each label and the joint size of the basin of attraction of each label."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>states</th><th scope=col>basin</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>IL10+</th><td>544, 608, 672, 736</td><td>178</td></tr>\n",
       "\t<tr><th scope=row>IL10+TGFB+</th><td>48, 112, 176, 240, 304, 368, 432, 496, 560, 624, 688, 752, 816, 880, 944, 1008</td><td>371</td></tr>\n",
       "\t<tr><th scope=row>IL10+TGFB+/TregIL10+</th><td>113/120, 241/248, 369/376, 497/504, 625/632, 753/760, 881/888, 1009/1016</td><td>67</td></tr>\n",
       "\t<tr><th scope=row>RORGT+/TGFB+</th><td>338/341, 466/469</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>TGFB+</th><td>16, 272</td><td>39</td></tr>\n",
       "\t<tr><th scope=row>Th0</th><td>0, 1, 4, 65, 68, 132, 129, 193, 196</td><td>152</td></tr>\n",
       "\t<tr><th scope=row>Th0/Th0</th><td>192/197</td><td>36</td></tr>\n",
       "\t<tr><th scope=row>Th17</th><td>278, 342, 406, 470</td><td>70</td></tr>\n",
       "\t<tr><th scope=row>Treg</th><td>89, 217, 345, 473</td><td>61</td></tr>\n",
       "\t<tr><th scope=row>TregIL10+</th><td>121, 249, 377, 505, 633, 761, 889, 1017</td><td>24</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       "  & states & basin\\\\\n",
       "\\hline\n",
       "\tIL10+ & 544, 608, 672, 736 & 178\\\\\n",
       "\tIL10+TGFB+ & 48, 112, 176, 240, 304, 368, 432, 496, 560, 624, 688, 752, 816, 880, 944, 1008 & 371\\\\\n",
       "\tIL10+TGFB+/TregIL10+ & 113/120, 241/248, 369/376, 497/504, 625/632, 753/760, 881/888, 1009/1016 & 67\\\\\n",
       "\tRORGT+/TGFB+ & 338/341, 466/469 & 26\\\\\n",
       "\tTGFB+ & 16, 272 & 39\\\\\n",
       "\tTh0 & 0, 1, 4, 65, 68, 132, 129, 193, 196 & 152\\\\\n",
       "\tTh0/Th0 & 192/197 & 36\\\\\n",
       "\tTh17 & 278, 342, 406, 470 & 70\\\\\n",
       "\tTreg & 89, 217, 345, 473 & 61\\\\\n",
       "\tTregIL10+ & 121, 249, 377, 505, 633, 761, 889, 1017 & 24\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "                                                                                             states\n",
       "IL10+                                                                            544, 608, 672, 736\n",
       "IL10+TGFB+           48, 112, 176, 240, 304, 368, 432, 496, 560, 624, 688, 752, 816, 880, 944, 1008\n",
       "IL10+TGFB+/TregIL10+       113/120, 241/248, 369/376, 497/504, 625/632, 753/760, 881/888, 1009/1016\n",
       "RORGT+/TGFB+                                                                       338/341, 466/469\n",
       "TGFB+                                                                                       16, 272\n",
       "Th0                                                             0, 1, 4, 65, 68, 132, 129, 193, 196\n",
       "Th0/Th0                                                                                     192/197\n",
       "Th17                                                                             278, 342, 406, 470\n",
       "Treg                                                                              89, 217, 345, 473\n",
       "TregIL10+                                                   121, 249, 377, 505, 633, 761, 889, 1017\n",
       "                     basin\n",
       "IL10+                  178\n",
       "IL10+TGFB+             371\n",
       "IL10+TGFB+/TregIL10+    67\n",
       "RORGT+/TGFB+            26\n",
       "TGFB+                   39\n",
       "Th0                    152\n",
       "Th0/Th0                 36\n",
       "Th17                    70\n",
       "Treg                    61\n",
       "TregIL10+               24"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attractors.by.label <- data.frame(\n",
    "    states = tapply(attr.df$involvedStates, attr.df$label, paste),\n",
    "    basin = tapply(attr.df$basin, attr.df$label, sum)\n",
    ")\n",
    "attractors.by.label"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This ends the basic BoolNet tutorial."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Next\n",
    "\n",
    "* [Introduction](./RPRN-Introduction.ipynb)\n",
    "* BoolNet\n",
    "* [Functions](./RPRN-Functions.ipynb)\n",
    "* [Updating](./RPRN-Updating.ipynb)\n",
    "* [States](./RPRN-States.ipynb)\n",
    "* [Appendix](./RPRN-Appendix.ipynb)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
